<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Seven Bridges Platform API Client and Common Workflow Language Tool Builder in R &bull; sevenbridges</title><!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous"><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet"><script src="jquery.sticky-kit.min.js"></script><script src="pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">sevenbridges</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="articles/index.html">Articles</a>
</li>
<li>
  <a href="news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/sbg/sevenbridges-r">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
    <div id="sevenbridges" class="section level1">
<div class="page-header"><h1 class="hasAnchor"><a href="#sevenbridges" class="anchor"> </a>sevenbridges <a href="http://www.sevenbridges.com"><img src="logo.png" align="right" alt="logo" height="128" width="128"></a></h1></div>

<p><a href="https://www.bioconductor.org/packages/release/bioc/html/sevenbridges.html">BioC (Release)</a> &middot; <a href="https://www.bioconductor.org/packages/devel/bioc/html/sevenbridges.html">BioC (Development)</a> &middot; <a href="https://github.com/sbg/sevenbridges-r">GitHub (Latest)</a></p>
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"> </a>Overview</h2>
<p>sevenbridges is an <a href="https://cran.r-project.org/">R</a>/<a href="https://www.bioconductor.org/">Bioconductor</a> package that provides an interface for the <a href="https://www.sbgenomics.com/">Seven Bridges Platform</a> and <a href="https://www.cancergenomicscloud.org/">Cancer Genomics Cloud</a> public APIs.</p>
<p>The Seven Bridges Platform is a cloud-based environment for conducting bioinformatic analyses. It is a central hub for teams to store, analyze, and jointly interpret their bioinformatic data. The Platform co-locates analysis workflows alongside the largest genomic datasets to optimize processing, allocating storage and compute resources on demand.</p>
<p>The Cancer Genomics Cloud (CGC), powered by Seven Bridges, is also a cloud-based computation environment. It was built as one of three pilot systems funded by the <a href="https://www.cancer.gov/">National Cancer Institute</a> to explore the paradigm of colocalizing massive genomics datasets, like <a href="https://cancergenome.nih.gov/">The Cancer Genome Atlas (TCGA)</a>, alongside secure and scalable computational resources to analyze them. The CGC makes more than a petabyte of multi-dimensional data available immediately to authorized researchers. You can add your own data to analyze alongside TCGA using predefined analytical workflows or your own tools.</p>
</div>
<div id="table-of-contents" class="section level2">
<h2 class="hasAnchor">
<a href="#table-of-contents" class="anchor"> </a>Table of Contents</h2>
<ul><li><a href="#overview">Overview</a></li>
<li><a href="#installation">Installation</a></li>
<li><a href="#bioconductor---release-version">Bioconductor - Release Version</a></li>
<li><a href="#bioconductor---development-version">Bioconductor - Development Version</a></li>
<li><a href="#latest-development-version">Latest Development Version</a></li>
<li><a href="#features">Features</a></li>
<li><a href="#flexible-authentication-methods">Flexible Authentication Methods</a></li>
<li><a href="#complete-api-r-client">Complete API R Client</a></li>
<li><a href="#task-monitoring">Task Monitoring</a></li>
<li><a href="#batch-tasks-support">Batch Tasks Support</a></li>
<li><a href="#cross-environment-support">Cross Environment Support</a></li>
<li><a href="#common-workflow-language-tool-interface">Common Workflow Language Tool Interface</a></li>
<li><a href="#utilities-for-tool-and-flow">Utilities for Tool and Flow</a></li>
<li><a href="#tutorials">Tutorials</a></li>
<li><a href="#ide-docker-image">IDE Docker Image</a></li>
<li><a href="#faq">FAQ</a></li>
<li><a href="#events">Events</a></li>
<li><a href="#contribute">Contribute</a></li>
</ul></div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"> </a>Installation</h2>
<div id="bioconductor---release-version" class="section level3">
<h3 class="hasAnchor">
<a href="#bioconductor---release-version" class="anchor"> </a>Bioconductor - Release Version</h3>
<p>This installation is recommended for most users as it is the most stable. The current release of Bioconductor is version 3.4, and it works with <strong>R version 3.3.2</strong>. Users of older R and Bioconductor versions should update their installation to take advantage of new features.</p>
<p>If you do not want to update R, please install the sevenbridges package directly from GitHub by following the instructions below.</p>
<div id="check-r-version" class="section level4">
<h4 class="hasAnchor">
<a href="#check-r-version" class="anchor"> </a>Check R version</h4>
<p>First, check which version of R you are using with the following:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">R.version.string</code></pre></div>
<p>If you are not running latest R, install R 3.3 by following <a href="https://www.bioconductor.org/install/#install-R">these instructions</a>. If you are using RStudio, close and restart RStudio after installing R 3.3. RStudio will detect the new installation.</p>
</div>
<div id="install-the-package" class="section level4">
<h4 class="hasAnchor">
<a href="#install-the-package" class="anchor"> </a>Install the package</h4>
<p>Install the sevenbridges package as shown below:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">"https://bioconductor.org/biocLite.R"</span>)
<span class="kw">biocLite</span>(<span class="st">"sevenbridges"</span>)</code></pre></div>
</div>
</div>
<div id="bioconductor---development-version" class="section level3">
<h3 class="hasAnchor">
<a href="#bioconductor---development-version" class="anchor"> </a>Bioconductor - Development Version</h3>
<p>If you are developing tools in the <code>devel</code> branch or if you are a user who uses the development version for R and Bioconductor, install the package from the bioconductor <code>devel</code> branch. You need to install R-devel first by following the directions in <a href="https://bioconductor.org/developers/how-to/useDevel/">&ldquo;Using the &lsquo;Devel&rsquo; Version of Bioconductor&rdquo;</a>. After upgrading R, this method of installation is harder. To try the latest features, please install directly from GitHub as next shown below:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">"https://bioconductor.org/biocLite.R"</span>)
<span class="kw">useDevel</span>(<span class="dt">devel =</span> <span class="ot">TRUE</span>)
<span class="kw">biocLite</span>(<span class="st">"sevenbridges"</span>)</code></pre></div>
<p>To load the package in R, simply call</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"sevenbridges"</span>)</code></pre></div>
</div>
<div id="latest-development-version" class="section level3">
<h3 class="hasAnchor">
<a href="#latest-development-version" class="anchor"> </a>Latest Development Version</h3>
<p>You can always install the latest development version of the sevenbridges package from GitHub. We push to the Bioconductor branch (<code>release</code>/<code>devel</code>) regularly.</p>
<div id="if-devtools-was-not-installed" class="section level4">
<h4 class="hasAnchor">
<a href="#if-devtools-was-not-installed" class="anchor"> </a>If <code>devtools</code> was not installed</h4>
<p>Installing the sevenbridges package requires you have the <code>devtools</code> package. If you do not have <code>devtools</code>, install it from <a href="https://cran.r-project.org/web/packages/devtools/README.html">CRAN</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">"devtools"</span>)</code></pre></div>
<p>You may got an error and need system dependecies for curl and ssl. For example, you probably need to do the following first in order to install <code>devtools</code> and in order to build vignettes since you need <code>pandoc</code> under Ubuntu.</p>
<pre><code>apt-get update
apt-get install libcurl4-gnutls-dev libssl-dev pandoc pandoc-citeproc</code></pre>
</div>
<div id="if-devtools-is-already-installed" class="section level4">
<h4 class="hasAnchor">
<a href="#if-devtools-is-already-installed" class="anchor"> </a>If <code>devtools</code> is already installed</h4>
<p>Install latest version of sevenbridges from GitHub with the following:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">"https://bioconductor.org/biocLite.R"</span>)
<span class="kw">biocLite</span>(<span class="st">"readr"</span>)

devtools::<span class="kw">install_github</span>(<span class="st">"sbg/sevenbridges-r"</span>, <span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>,
                         <span class="dt">repos =</span> BiocInstaller::<span class="kw">biocinstallRepos</span>(),
                         <span class="dt">dependencies =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>If you have trouble with <code>pandoc</code> and do not want to install it, set <code>build_vignettes = FALSE</code> to avoid building the vignettes.</p>
</div>
</div>
</div>
<div id="features" class="section level2">
<h2 class="hasAnchor">
<a href="#features" class="anchor"> </a>Features</h2>
<p>The sevenbridges package includes the following features:</p>
<div id="flexible-authentication-methods" class="section level3">
<h3 class="hasAnchor">
<a href="#flexible-authentication-methods" class="anchor"> </a>Flexible Authentication Methods</h3>
<ul><li>Multiple authentication methods support.</li>
</ul><p>Direct authentication:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># direct authentication</span>
a &lt;-<span class="st"> </span><span class="kw"><a href="reference/Auth-class.html">Auth</a></span>(<span class="dt">token =</span> <span class="st">"your_token"</span>, <span class="dt">platform =</span> <span class="st">"cgc"</span>)
<span class="co"># or use base url</span>
a &lt;-<span class="st"> </span><span class="kw"><a href="reference/Auth-class.html">Auth</a></span>(<span class="dt">token =</span> <span class="st">"your_token"</span>, <span class="dt">url =</span> <span class="st">"https://cgc-api.sbgenomics.com/v2"</span>)</code></pre></div>
<p>Authentication via system environment variables:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="reference/sbg_set_env.html">sbg_set_env</a></span>(<span class="dt">token =</span> <span class="st">"your_token"</span>, <span class="dt">url =</span> <span class="st">"https://cgc-api.sbgenomics.com/v2"</span>)
a &lt;-<span class="st"> </span><span class="kw"><a href="reference/Auth-class.html">Auth</a></span>(<span class="dt">from =</span> <span class="st">"env"</span>)</code></pre></div>
<p>Authentication via a user configuration file, collect and manage your credentials for multiple accounts across various Seven Bridges environments:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a &lt;-<span class="st"> </span><span class="kw"><a href="reference/Auth-class.html">Auth</a></span>(<span class="dt">from =</span> <span class="st">"file"</span>, <span class="dt">profile_name =</span> <span class="st">"aws-us-tengfei"</span>)</code></pre></div>
<p>Please check <code><a href="articles/api.html">vignette("api", package = "sevenbridges")</a></code> for technical details about all available authentication methods.</p>
</div>
<div id="complete-api-r-client" class="section level3">
<h3 class="hasAnchor">
<a href="#complete-api-r-client" class="anchor"> </a>Complete API R Client</h3>
<ul><li>A complete API R client with a user-friendly, object-oriented API with printing and support operations for API requests relating to users, billing, projects, files, apps, and tasks. Short examples are also included, as shown below:</li>
</ul><div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Get a project by pattern-matching its name</span>
p &lt;-<span class="st"> </span>a$<span class="kw">project</span>(<span class="st">"demo"</span>)
<span class="co"># Get a project by its id</span>
p &lt;-<span class="st"> </span>a$<span class="kw">project</span>(<span class="dt">id =</span> <span class="st">"tengfei/demo"</span>)
<span class="co"># Delete files from a project</span>
p$<span class="kw">file</span>(<span class="st">"sample.tz"</span>)$<span class="kw"><a href="reference/delete.html">delete</a></span>()
<span class="co"># Upload fies from a folder to a project and include file metadata</span>
p$<span class="kw">upload</span>(<span class="st">"folder_path"</span>, <span class="dt">metadata =</span> <span class="kw">list</span>(<span class="dt">platform =</span> <span class="st">"Illumina"</span>))</code></pre></div>
</div>
<div id="task-monitoring" class="section level3">
<h3 class="hasAnchor">
<a href="#task-monitoring" class="anchor"> </a>Task Monitoring</h3>
<ul><li>A task monitoring hook which allows you to add a hook function to specific task statuses as you monitor a task. For example, you can opt to receive an email when the task is completed or specify to download all files produced by the task, as shown below:</li>
</ul><div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="reference/setTaskHook.html">setTaskHook</a></span>(<span class="st">"completed"</span>, function() {
    tsk$<span class="kw"><a href="reference/download.html">download</a></span>(<span class="st">"~/Downloads"</span>)
})
tsk$<span class="kw">monitor</span>()</code></pre></div>
</div>
<div id="batch-tasks-support" class="section level3">
<h3 class="hasAnchor">
<a href="#batch-tasks-support" class="anchor"> </a>Batch Tasks Support</h3>
<ul><li>Batch tasks by metadata and by item.</li>
</ul><div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Batch by item</span>
(tsk &lt;-<span class="st"> </span>p$<span class="kw">task_add</span>(<span class="dt">name =</span> <span class="st">"RNA DE report new batch 2"</span>,
                   <span class="dt">description =</span> <span class="st">"RNA DE analysis report"</span>,
                   <span class="dt">app =</span> rna.app$id,
                   <span class="dt">batch =</span> <span class="kw"><a href="reference/batch.html">batch</a></span>(<span class="dt">input =</span> <span class="st">"bamfiles"</span>),
                   <span class="dt">inputs =</span> <span class="kw">list</span>(<span class="dt">bamfiles =</span> bamfiles.in,
                                 <span class="dt">design =</span> design.in,
                                 <span class="dt">gtffile =</span> gtf.in)))

<span class="co"># Batch by metadata. Note that input files must have relevant metadata fields specified.</span>
(tsk &lt;-<span class="st"> </span>p$<span class="kw">task_add</span>(<span class="dt">name =</span> <span class="st">"RNA DE report new batch 3"</span>,
                   <span class="dt">description =</span> <span class="st">"RNA DE analysis report"</span>,
                   <span class="dt">app =</span> rna.app$id,
                   <span class="dt">batch =</span> <span class="kw"><a href="reference/batch.html">batch</a></span>(<span class="dt">input =</span> <span class="st">"fastq"</span>,
                                 <span class="kw">c</span>(<span class="st">"metadata.sample_id"</span>,
                                 <span class="st">"metadata.library_id"</span>)),
                   <span class="dt">inputs =</span> <span class="kw">list</span>(<span class="dt">bamfiles =</span> bamfiles.in,
                                 <span class="dt">design =</span> design.in,
                                 <span class="dt">gtffile =</span> gtf.in)))</code></pre></div>
</div>
<div id="cross-environment-support" class="section level3">
<h3 class="hasAnchor">
<a href="#cross-environment-support" class="anchor"> </a>Cross Environment Support</h3>
<ul><li>Cross-platform support for Seven Bridges environments, such as <a href="https://www.cancergenomicscloud.org/">Cancer Genomics Cloud</a> or <a href="https://www.sbgenomics.com/">Seven Bridges Platform</a> on either Amazon Web Services or Google Cloud Platform.</li>
</ul></div>
<div id="common-workflow-language-tool-interface" class="section level3">
<h3 class="hasAnchor">
<a href="#common-workflow-language-tool-interface" class="anchor"> </a>Common Workflow Language Tool Interface</h3>
<ul><li>A <a href="http://www.commonwl.org/">Common Workflow Language (CWL)</a> Tool interface to directly describe your tool in R, export it to JSON or YAML, or add it to your online project. This package defines a complete set of CWL object, so you can describe tools as follows:</li>
</ul><div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"readr"</span>)
fd &lt;-<span class="st"> </span><span class="kw"><a href="reference/CPURequirement-class.html">fileDef</a></span>(<span class="dt">name =</span> <span class="st">"runif.R"</span>,
              <span class="dt">content =</span> <span class="kw">read_file</span>(fl))
rbx &lt;-<span class="st"> </span><span class="kw"><a href="reference/Tool-class.html">Tool</a></span>(<span class="dt">id =</span> <span class="st">"runif"</span>,
            <span class="dt">label =</span> <span class="st">"runif"</span>,
            <span class="dt">hints =</span> <span class="kw"><a href="reference/CPURequirement-class.html">requirements</a></span>(<span class="kw"><a href="reference/CPURequirement-class.html">docker</a></span>(<span class="dt">pull =</span> <span class="st">"rocker/r-base"</span>),
                                 <span class="kw"><a href="reference/CPURequirement-class.html">cpu</a></span>(<span class="dv">1</span>), <span class="kw"><a href="reference/CPURequirement-class.html">mem</a></span>(<span class="dv">2000</span>)),
            <span class="dt">requirements =</span> <span class="kw"><a href="reference/CPURequirement-class.html">requirements</a></span>(fd),
            <span class="dt">baseCommand =</span> <span class="st">"Rscript runif.R"</span>,
            <span class="dt">stdout =</span> <span class="st">"output.txt"</span>,
            <span class="dt">inputs =</span> <span class="kw">list</span>(<span class="kw"><a href="reference/CLB.html">input</a></span>(<span class="dt">id =</span> <span class="st">"number"</span>,
                                <span class="dt">type =</span> <span class="st">"integer"</span>,
                                <span class="dt">position =</span> <span class="dv">1</span>),
                          <span class="kw"><a href="reference/CLB.html">input</a></span>(<span class="dt">id =</span> <span class="st">"min"</span>,
                                <span class="dt">type =</span> <span class="st">"float"</span>,
                                <span class="dt">position =</span> <span class="dv">2</span>),
                          <span class="kw"><a href="reference/CLB.html">input</a></span>(<span class="dt">id =</span> <span class="st">"max"</span>,
                                <span class="dt">type =</span> <span class="st">"float"</span>,
                                <span class="dt">position =</span> <span class="dv">3</span>)),
            <span class="dt">outputs =</span> <span class="kw"><a href="reference/CLB.html">output</a></span>(<span class="dt">id =</span> <span class="st">"random"</span>, <span class="dt">glob =</span> <span class="st">"output.txt"</span>))
<span class="co"># output CWL JSON</span>
rbx$<span class="kw">toJSON</span>(<span class="dt">pretty =</span> <span class="ot">TRUE</span>)
<span class="co"># output CWL YAML</span>
rbx$<span class="kw">toYAML</span>()</code></pre></div>
</div>
<div id="utilities-for-tool-and-flow" class="section level3">
<h3 class="hasAnchor">
<a href="#utilities-for-tool-and-flow" class="anchor"> </a>Utilities for Tool and Flow</h3>
<ul><li>Utilities for Tool and Flow, for example</li>
</ul><div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># converting a SBG CWL JSON file</span>
<span class="kw">library</span>(<span class="st">"sevenbridges"</span>)
t1 &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/app"</span>, <span class="st">"tool_star.json"</span>, <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
<span class="co"># convert json file into a Tool object</span>
t1 &lt;-<span class="st"> </span><span class="kw"><a href="reference/convert_app.html">convert_app</a></span>(t1)
<span class="co"># shows all input matrix</span>
t1$<span class="kw"><a href="reference/input_matrix.html">input_matrix</a></span>()</code></pre></div>
</div>
</div>
<div id="tutorials" class="section level2">
<h2 class="hasAnchor">
<a href="#tutorials" class="anchor"> </a>Tutorials</h2>
<p>We maintain 3 different sets of documentation: the sevenbridges-r GitHub repository (latest and most up-to-date), Bioconductor release channel, and Bioconductor development channel. Below, only the GitHub version is linked to provide the latest documentation. For the other versions, please visit <a href="https://www.bioconductor.org/packages/release/bioc/html/sevenbridges.html">Bioconductor Release version</a> or <a href="https://www.bioconductor.org/packages/devel/bioc/html/sevenbridges.html">Bioconductor Development version</a>. The tutorials below are re-generated regularly as we update the package on GitHub.</p>
<table class="table table-bordered table-striped" style="width:100%;"><colgroup><col width="70%"><col width="15%"><col width="15%"></colgroup><thead><tr class="header"><th align="left">Tutorial Title</th>
<th align="left">HTML</th>
<th align="left">Rmd Source</th>
</tr></thead><tbody><tr class="odd"><td align="left">Complete Reference for the API R Client</td>
<td align="left"><a href="https://sbg.github.io/sevenbridges-r/articles/api.html">HTML</a></td>
<td align="left"><a href="https://raw.githubusercontent.com/sbg/sevenbridges-r/master/vignettes/api.Rmd">Source</a></td>
</tr><tr class="even"><td align="left">Use R on the Cancer Genomics Cloud</td>
<td align="left"><a href="https://sbg.github.io/sevenbridges-r/articles/bioc-workflow.html">HTML</a></td>
<td align="left"><a href="https://raw.githubusercontent.com/sbg/sevenbridges-r/master/vignettes/bioc-workflow.Rmd">Source</a></td>
</tr><tr class="odd"><td align="left">Create a Docker Container and use Command Line Interface for R</td>
<td align="left"><a href="https://sbg.github.io/sevenbridges-r/articles/docker.html">HTML</a></td>
<td align="left"><a href="https://raw.githubusercontent.com/sbg/sevenbridges-r/master/vignettes/docker.Rmd">Source</a></td>
</tr><tr class="even"><td align="left">Describe and execute Common Workflow Language (CWL) Tools and Workflows in R</td>
<td align="left"><a href="https://sbg.github.io/sevenbridges-r/articles/apps.html">HTML</a></td>
<td align="left"><a href="https://raw.githubusercontent.com/sbg/sevenbridges-r/master/vignettes/apps.Rmd">Source</a></td>
</tr><tr class="odd"><td align="left">
<a name="ide"></a>IDE container: Rstudio and Shiny server and more</td>
<td align="left"><a href="https://sbg.github.io/sevenbridges-r/articles/rstudio.html">HTML</a></td>
<td align="left"><a href="https://raw.githubusercontent.com/sbg/sevenbridges-r/master/vignettes/rstudio.Rmd">Source</a></td>
</tr><tr class="even"><td align="left">Browse data on the Cancer Genomics Cloud via the Data Explorer, a SPARQL query, <br>or the Datasets API</td>
<td align="left"><a href="https://sbg.github.io/sevenbridges-r/articles/cgc-sparql.html">HTML</a></td>
<td align="left"><a href="https://raw.githubusercontent.com/sbg/sevenbridges-r/master/vignettes/cgc-sparql.Rmd">Source</a></td>
</tr></tbody></table></div>
<div id="ide-docker-image" class="section level2">
<h2 class="hasAnchor">
<a href="#ide-docker-image" class="anchor"> </a>IDE Docker Image</h2>
<p>In the tutorial for <a href="#tutorials">IDE container</a> above, we built a Docker container locally from which we can launch RStudio and Shiny. To launch RStudio and Shiny Server with the Seven Bridges IDE Docker container, do the following:</p>
<pre class="shell"><code>docker run  -d -p 8787:8787 -p 3838:3838 --name rstudio_shiny_server sevenbridges/sevenbridges-r</code></pre>
<p>To mount a file system, you need to use <code>--privileged</code> with fuse.</p>
<pre class="shell"><code>docker run  --privileged -d -p 8787:8787 -p 3838:3838 --name rstudio_shiny_server sevenbridges/sevenbridges-r</code></pre>
<p>Check out the IP from Docker Machine if you are on a Mac OS.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">docker-machine</span> ip default</code></pre></div>
<p>In your browser, you can see where the RStudio server is located from the path <code>http://&lt;url&gt;:8787/</code>. For example, if 192.168.99.100 is returned, visit <code>http://192.168.99.100:8787/</code> for Rstudio.</p>
<p>For the Shiny server, each app__ is hosted at <code>http://&lt;url&gt;:3838/users/&lt;username of rstudio&gt;/&lt;app_dir&gt;</code> for the Shiny server. For example, an app called <code>01_hello</code> owned by user <code>rstudio</code> (a default user) has the path <code>http://&lt;url&gt;:3838/users/rstudio/01_hello/</code>. To develop your Shiny apps as an Rstudio user, you can login your RStudio server and create a folder in your home folder called <code>~/ShinyApps</code>. There, you can develop shiny apps in that folder. For example, you can create an app called <code>02_text</code> at <code>~/ShinyApps/02_text/</code>.</p>
<p>Log into your RStudio at <code>http://&lt;url&gt;:8787</code>. Then, try to copy an app to your home folder, as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dir.create</span>(<span class="st">"~/ShinyApps"</span>)
<span class="kw">file.copy</span>(<span class="st">"/usr/local/lib/R/site-library/shiny/examples/01_hello/"</span>,
          <span class="st">"~/ShinyApps/"</span>, <span class="dt">recursive =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>If you are logged in as user <code>rstudio</code>, visit <code>http://192.168.99.100:3838/rstudio/01_hello</code>. You should be able to see the &ldquo;hello&rdquo; example.</p>
<p><em>Note</em>: Generic Shiny apps can also be hosted at <code>http://&lt;url&gt;:3838/</code> or, for a particular app, at <code>http://&lt;url&gt;:3838/&lt;app_dir&gt;</code>. Inside the Docker container, it&rsquo;s hosted under <code>/srv/shiny-server/</code>.</p>
</div>
<div id="faq" class="section level2">
<h2 class="hasAnchor">
<a href="#faq" class="anchor"> </a>FAQ</h2>
<p>The best place to ask questions about the sevenbridges package is the <a href="https://groups.google.com/forum/#!forum/sevenbridges-r">mailing list</a>.</p>
<ul><li><p><strong>Q</strong>: Does this package support Seven Bridges&rsquo; API v1 which was not CWL compatible?<br><strong>A</strong>: No. This package only supports API v2 +. For API v1, please check out the <a href="https://github.com/road2stat/sbgr">sbgr</a> package. Note that API v1 and associated legacy project types will be deprecated.</p></li>
<li><p><strong>Q</strong>: Which version of the Common Workflow Language (CWL) is supported?<br><strong>A</strong>: We support draft 2 and are making progress on supporting draft 3.</p></li>
<li><p><strong>Q</strong>: Is there a Python binding for the API?<br><strong>A</strong>: Yes, the official Python client is <a href="https://github.com/sbg/sevenbridges-python">here</a>. Recipes and tutorials using the Python bindings are <a href="https://github.com/sbg/okAPI">here</a>.</p></li>
<li><p><strong>Q</strong>: Why do I get warning messages when I use the API R client?<br><strong>A</strong>: The warning only exists in Rstudio and is potentially a bug in Rstudio. To ignore, it use <code>options(warn = -1)</code></p></li>
<li><p><strong>Q</strong>: I still have problems despite dismissing the messages.<br><strong>A</strong>: Please try to use the latest package on GitHub or update your package on Bioconductor. This usually includes the most recent bug fixes.</p></li>
</ul></div>
<div id="events" class="section level2">
<h2 class="hasAnchor">
<a href="#events" class="anchor"> </a>Events</h2>
<table class="table table-bordered table-striped" style="width:100%;"><colgroup><col width="20%"><col width="45%"><col width="35%"></colgroup><thead><tr class="header"><th align="left">Time</th>
<th align="left">Event</th>
<th align="left">Location</th>
</tr></thead><tbody><tr class="odd"><td align="left">Jan 12, 2017</td>
<td align="left">
<a href="https://www.meetup.com/Boston-R-Bioconductor-for-genomics/events/235580582/">Genomics in the Cloud - Boston Bioconductor Meetup</a> (talk)</td>
<td align="left">Dana-Farber Cancer Institute, Boston, MA</td>
</tr><tr class="even"><td align="left">Sep 12 - 14, 2016</td>
<td align="left">
<a href="https://www.stats.ox.ac.uk/events/probgen16">Probabilistic Modeling in Genomics</a> (poster)</td>
<td align="left">University of Oxford, Oxford, UK</td>
</tr><tr class="odd"><td align="left">May 27 - 29, 2016</td>
<td align="left">
<a href="http://china-r.org/bj2016/index.html">The 9th China-R Conference</a> (talk)</td>
<td align="left">Renmin University of China, Beijng, China</td>
</tr><tr class="even"><td align="left">Jun 27 - 30, 2016</td>
<td align="left">
<a href="http://user2016.org/">The R User Conference 2016</a> (talk)</td>
<td align="left">Stanford University, Stanford, CA</td>
</tr><tr class="odd"><td align="left">Jun 24 - 26, 2016</td>
<td align="left">
<a href="https://bioconductor.org/help/course-materials/2016/BioC2016/">BioC 2016: Where Software and Biology Connect</a> (workshop)</td>
<td align="left">Stanford University, Stanford, CA</td>
</tr><tr class="even"><td align="left">Apr 1 - 3, 2016</td>
<td align="left">
<a href="https://www.cancergenomicscloud.org/hacking-cancer/">NCI Cancer Genomics Cloud Hackathon</a> (tutorial)<br>[<a href="https://sbg.github.io/sevenbridges-r/articles/bioc-workflow.html">HTML</a>] [<a href="https://raw.githubusercontent.com/sbg/sevenbridges-r/master/vignettes/bioc-workflow.Rmd">R Markdown Source</a>]</td>
<td align="left">Seven Bridges Genomics, Inc., Boston, MA</td>
</tr></tbody></table></div>
<div id="contribute" class="section level2">
<h2 class="hasAnchor">
<a href="#contribute" class="anchor"> </a>Contribute</h2>
<p>All feedback is welcome! Please file bug reports/feature requests on the <a href="https://github.com/sbg/sevenbridges-r/issues">issue page</a>, or create pull requests <a href="https://github.com/sbg/sevenbridges-r/pulls">here</a>.</p>
<hr><p>&copy; 2016 Seven Bridges Genomics, Inc. Licensed under the Apache License 2.0.</p>
</div>
</div>
  </div>

  <div class="col-md-3" id="sidebar">
    <h2>Links</h2><ul class="list-unstyled"><li>Browse source code at <br><a href="https://github.com/sbg/sevenbridges-r">https://&#8203;github.com/&#8203;sbg/&#8203;sevenbridges-r</a></li>
<li>Report a bug at <br><a href="https://github.com/sbg/sevenbridges-r/issues">https://&#8203;github.com/&#8203;sbg/&#8203;sevenbridges-r/&#8203;issues</a></li>
</ul><h2>License</h2>
<p>Apache License 2.0 | file <a href="LICENSE">LICENSE</a></p>
<h2>Developers</h2><ul class="list-unstyled"><li>Nan Xiao <br><small class="roles"> Author, maintainer </small> </li>
<li>Dusan Randjelovic <br><small class="roles"> Author </small> </li>
<li>Tengfei Yin <br><small class="roles"> Author </small> </li>
<li>Seven Bridges Genomics <br><small class="roles"> Copyright&nbsp;holder, funder </small> </li>
<li><a href="authors.html">All authors...</a></li>
</ul><html><body><h2>Dev status</h2><ul class="list-unstyled"><li><a href="https://travis-ci.org/sbg/sevenbridges-r">
  <img src="https://travis-ci.org/sbg/sevenbridges-r.svg?branch=master" alt="Build Status"></a></li>
<li><a href="https://ci.appveyor.com/project/road2stat/sevenbridges-r">
  <img src="https://ci.appveyor.com/api/projects/status/github/sbg/sevenbridges-r?branch=master&amp;svg=true" alt="AppVeyor Build Status"></a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/sevenbridges.html#archives">
  <img src="https://www.bioconductor.org/shields/availability/devel/sevenbridges.svg" alt="platform"></a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/sevenbridges.html#since">
  <img src="https://www.bioconductor.org/shields/years-in-bioc/sevenbridges.svg" alt="bioc"></a></li>
<li><a href="https://bioconductor.org/packages/stats/bioc/sevenbridges/">
  <img src="https://www.bioconductor.org/shields/downloads/sevenbridges.svg" alt="downloads"></a></li>
</ul></body></html></div>
</div>


      <footer><div class="copyright">
  <p>Developed by Nan Xiao, Dusan Randjelovic, Tengfei Yin, Seven Bridges Genomics.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer></div>

<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-87185344-1', 'auto');
  ga('send', 'pageview');

</script>
  </body></html>
