<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Describe and Execute CWL Tools/Workflows in R • sevenbridges</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Google analytics --><script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-87185344-1', 'auto');
  ga('send', 'pageview');

</script>
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">sevenbridges-r</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/api.html">Complete Guide for Seven Bridges API R Client</a>
    </li>
    <li>
      <a href="../articles/apps.html">Describe and Execute CWL Tools/Workflows in R</a>
    </li>
    <li>
      <a href="../articles/bioc-workflow.html">Master Tutorial: Use R for Cancer Genomics Cloud</a>
    </li>
    <li>
      <a href="../articles/cgc-sparql.html">Find Data on CGC via Data Exploerer, SPARQL, and Datasets API</a>
    </li>
    <li>
      <a href="../articles/docker.html">Creating Your Docker Container and Command Line Interface (with docopt)</a>
    </li>
    <li>
      <a href="../articles/rstudio.html">IDE Container: RStudio Server, Shiny Server, and More</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/sbg/sevenbridges-r">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Describe and Execute CWL Tools/Workflows in R</h1>
                        <h4 class="author">Tengfei Yin &lt;<a href="mailto:tengfei.yin@sevenbridges.com">tengfei.yin@sevenbridges.com</a>&gt;</h4>
            
            <h4 class="date">2018-02-14</h4>
          </div>

    
    
<div class="contents">
<div id="prerequisite" class="section level1">
<h1 class="hasAnchor">
<a href="#prerequisite" class="anchor"></a>Prerequisite</h1>
<p>This tutorial assume you have basic knowledge about Docker concept.</p>
<p>Note: Right now we are supporting CWL draft 2 with SBG extension, but we will support CWL v1.0 soon.</p>
</div>
<div id="app-workflow-and-tool" class="section level1">
<h1 class="hasAnchor">
<a href="#app-workflow-and-tool" class="anchor"></a>App, Workflow, and Tool</h1>
<p>In our terminology, a <strong>workflow</strong> is composed of one or more <strong>tool</strong>, both of them are just <strong>app</strong> to users. You can imagine some raw input data go through a pipeline with many nodes that each step perform a function on the data in the flow, and in the end, you got want you want: a fully processed data or result (plot, report, action).</p>
<p>Here are some key ideas:</p>
<ul>
<li>Tool is the unit or a single node of workflow, so different tools could be connected into a workflow. That’s how we achieve reusability of components.</li>
<li>Tool is described with key components: input, output, parameters and requirements and more details. You understand the tool more like a black box (container) which digest some input(s) with specified setup and output another format.</li>
<li>When input files matches output between two tools, they could be connected.</li>
<li>Input is composed of files and parameters, we call it types: File, Enum, Integer, String and so on.</li>
<li>App could be described in JSON/YAML format, following Common Workflow Language (CWL) open-source standard.</li>
<li>CWL is just the collection of logic and schema. To execute this <strong>pure text file</strong>, we need executor in the cloud or local. With Seven Bridges platform, you can simply execute it at scale.</li>
</ul>
<p>Looks like full of jargons and hard to understand. Here is an example. You have a csv table, full of missing value and you want to process it in 3 steps:</p>
<ol style="list-style-type: decimal">
<li>Replace missing value</li>
<li>Filtering out rows that column “age” is smaller than 10</li>
<li>Output 3 item: a processed table csv file, a plot and a summary report in PDF format.</li>
</ol>
<p>You can describe each step into a single module or tool then connect them one by one to form a flow. You can put everything into one single “tool”, then downside is that other user cannot use your step1 for missing value problem. So it’s both art and sciense to leverage between flexibility and efficiency.</p>
<p>Why we are using CWL? Imagine a single file represeting a tool or workflow, could be executed anywhere in a reproducible manner and you don’t have to install anything because Docker container is imaged, that’s going to change the world of computational scientific research and how we do research and publish results. In this package we are trying to hide CWL details as much as possible, so user can just use it like a typical R function.</p>
</div>
<div id="describe-tools-in-r" class="section level1">
<h1 class="hasAnchor">
<a href="#describe-tools-in-r" class="anchor"></a>Describe Tools in R</h1>
<p><code>Tool</code> is the basic unit, and also your “lego brick” you usually start with. As developer you also want to provide those “lego” piecies to users to directly run it or make their own flow with it.</p>
<p>The main interface provided by <code>sevenbridges</code> package is <code>Tool</code> function, it’s much more straight forward to describe than composing your raw CWL JSON file from scratch. A “Tool” object in R could be exported into JSON or imported from a CWL JSON file.</p>
<p>I highly recommend going over documentation <a href="http://docs.cancergenomicscloud.org/docs/the-tool-editor">The Tool Editor</a> chapter for the Cancer Genomics Cloud to understand how it works, and even try it on the platform with the GUI. This will help understand our R interface better.</p>
<div id="import-from-json-file" class="section level2">
<h2 class="hasAnchor">
<a href="#import-from-json-file" class="anchor"></a>Import from JSON file</h2>
<p>Sometimes people share Tool in pure JSON text format. You can simply load it into R by using <code>convert_app</code> function, this will recognize your JSON file class (Tool or Workflow) automatically.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"sevenbridges"</span>)
t1 =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/app"</span>, <span class="st">"tool_star.json"</span>, <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
<span class="co"># # convert json file into a Tool object</span>
t1 =<span class="st"> </span><span class="kw"><a href="../reference/convert_app.html">convert_app</a></span>(t1)
<span class="co"># # try print it yourself</span>
<span class="co"># t1</span></code></pre></div>
<p>In this way, you can load it, revise it, use it with API or edit and export it back to JSON file. However, in this tutorial, the most important thing is that you learn how to desribe it directly in R.</p>
</div>
<div id="utilitites-for-tool-object" class="section level2">
<h2 class="hasAnchor">
<a href="#utilitites-for-tool-object" class="anchor"></a>Utilitites for Tool object</h2>
<p>We provide couple utitlities to help construct your own CWL tool quickly in R. For all availale utiles please check out <code>help("Tool")</code></p>
<p>Some utiles you will find it useful when you execute a task, you need to know what is the input type and what is the input id and if it’s required or not, so you can execute the task with parameters it need. Try play with <code>input_matrix</code> or <code>input_type</code> as shown below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get input type information</span>
<span class="kw">head</span>(t1<span class="op">$</span><span class="kw">input_type</span>())</code></pre></div>
<pre><code>              reads  readMatesLengthsIn       readMapNumber 
          "File..."              "enum"               "int" 
  limitOutSJoneRead limitOutSJcollapsed    outReadsUnmapped 
              "int"               "int"              "enum" </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get output type information</span>
<span class="kw">head</span>(t1<span class="op">$</span><span class="kw">output_type</span>())</code></pre></div>
<pre><code>              aligned_reads transcriptome_aligned_reads 
                     "File"                      "File" 
             reads_per_gene                   log_files 
                     "File"                   "File..." 
           splice_junctions          chimeric_junctions 
                     "File"                      "File" </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># return a input matrix with more informtion</span>
<span class="kw">head</span>(t1<span class="op">$</span><span class="kw"><a href="../reference/input_output_matrix.html">input_matrix</a></span>())</code></pre></div>
<pre><code>                     id                label    type required
1                #reads        Read sequence File...     TRUE
95         #sjdbGTFfile Splice junction file File...    FALSE
102             #genome         Genome files    File     TRUE
2   #readMatesLengthsIn        Reads lengths    enum    FALSE
3        #readMapNumber         Reads to map     int    FALSE
4    #limitOutSJoneRead Junctions max number     int    FALSE
                  prefix
1                   &lt;NA&gt;
95                  &lt;NA&gt;
102                 &lt;NA&gt;
2   --readMatesLengthsIn
3        --readMapNumber
4    --limitOutSJoneRead
                                                   fileTypes
1   FASTA, FASTQ, FA, FQ, FASTQ.GZ, FQ.GZ, FASTQ.BZ2, FQ.BZ2
95                                             GTF, GFF, TXT
102                                                      TAR
2                                                       null
3                                                       null
4                                                       null</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># return only a few fields</span>
<span class="kw">head</span>(t1<span class="op">$</span><span class="kw"><a href="../reference/input_output_matrix.html">input_matrix</a></span>(<span class="kw">c</span>(<span class="st">"id"</span>, <span class="st">"type"</span>, <span class="st">"required"</span>)))</code></pre></div>
<pre><code>                     id    type required
1                #reads File...     TRUE
95         #sjdbGTFfile File...    FALSE
102             #genome    File     TRUE
2   #readMatesLengthsIn    enum    FALSE
3        #readMapNumber     int    FALSE
4    #limitOutSJoneRead     int    FALSE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># return only required</span>
t1<span class="op">$</span><span class="kw"><a href="../reference/input_output_matrix.html">input_matrix</a></span>(<span class="dt">required =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>         id         label    type required prefix
1    #reads Read sequence File...     TRUE   &lt;NA&gt;
102 #genome  Genome files    File     TRUE   &lt;NA&gt;
                                                   fileTypes
1   FASTA, FASTQ, FA, FQ, FASTQ.GZ, FQ.GZ, FASTQ.BZ2, FQ.BZ2
102                                                      TAR</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># return a output matrix with more informtion</span>
t1<span class="op">$</span><span class="kw"><a href="../reference/input_output_matrix.html">output_matrix</a></span>()</code></pre></div>
<pre><code>                            id                     label    type fileTypes
1               #aligned_reads           Aligned SAM/BAM    File  SAM, BAM
2 #transcriptome_aligned_reads  Transcriptome alignments    File       BAM
3              #reads_per_gene            Reads per gene    File       TAB
4                   #log_files                 Log files File...       OUT
5            #splice_junctions          Splice junctions    File       TAB
6          #chimeric_junctions        Chimeric junctions    File  JUNCTION
7              #unmapped_reads            Unmapped reads File...     FASTQ
8         #intermediate_genome Intermediate genome files    File       TAR
9         #chimeric_alignments       Chimeric alignments    File       SAM</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># return only a few fields</span>
t1<span class="op">$</span><span class="kw"><a href="../reference/input_output_matrix.html">output_matrix</a></span>(<span class="kw">c</span>(<span class="st">"id"</span>, <span class="st">"type"</span>))</code></pre></div>
<pre><code>                            id    type
1               #aligned_reads    File
2 #transcriptome_aligned_reads    File
3              #reads_per_gene    File
4                   #log_files File...
5            #splice_junctions    File
6          #chimeric_junctions    File
7              #unmapped_reads File...
8         #intermediate_genome    File
9         #chimeric_alignments    File</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get required input id</span>
t1<span class="op">$</span><span class="kw">get_required</span>()</code></pre></div>
<pre><code>    reads    genome 
"File..."    "File" </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># set new required input with ID, # or without #</span>
t1<span class="op">$</span><span class="kw">set_required</span>(<span class="kw">c</span>(<span class="st">"#reads"</span>, <span class="st">"winFlankNbins"</span>))</code></pre></div>
<pre><code>not implemented yet!</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">t1<span class="op">$</span><span class="kw">get_required</span>()</code></pre></div>
<pre><code>    reads    genome 
"File..."    "File" </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># turn off requirements for input node #reads</span>
t1<span class="op">$</span><span class="kw">set_required</span>(<span class="st">"reads"</span>, <span class="ot">FALSE</span>)</code></pre></div>
<pre><code>not implemented yet!</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">t1<span class="op">$</span><span class="kw">get_required</span>()</code></pre></div>
<pre><code>    reads    genome 
"File..."    "File" </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get input id</span>
<span class="kw">head</span>(t1<span class="op">$</span><span class="kw">input_id</span>())</code></pre></div>
<pre><code>                 #STAR                  #STAR                  #STAR 
              "#reads"  "#readMatesLengthsIn"       "#readMapNumber" 
                 #STAR                  #STAR                  #STAR 
  "#limitOutSJoneRead" "#limitOutSJcollapsed"    "#outReadsUnmapped" </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get full input id with Tool name</span>
<span class="kw">head</span>(t1<span class="op">$</span><span class="kw">input_id</span>(<span class="ot">TRUE</span>))</code></pre></div>
<pre><code>                    File...                        enum 
              "#STAR.reads"  "#STAR.readMatesLengthsIn" 
                        int                         int 
      "#STAR.readMapNumber"   "#STAR.limitOutSJoneRead" 
                        int                        enum 
"#STAR.limitOutSJcollapsed"    "#STAR.outReadsUnmapped" </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get output id</span>
<span class="kw">head</span>(t1<span class="op">$</span><span class="kw">output_id</span>())</code></pre></div>
<pre><code>                         #STAR                          #STAR 
              "#aligned_reads" "#transcriptome_aligned_reads" 
                         #STAR                          #STAR 
             "#reads_per_gene"                   "#log_files" 
                         #STAR                          #STAR 
           "#splice_junctions"          "#chimeric_junctions" </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get full output id</span>
<span class="kw">head</span>(t1<span class="op">$</span><span class="kw">output_id</span>(<span class="ot">TRUE</span>))</code></pre></div>
<pre><code>                               File                                File 
              "#STAR.aligned_reads" "#STAR.transcriptome_aligned_reads" 
                               File                             File... 
             "#STAR.reads_per_gene"                   "#STAR.log_files" 
                               File                                File 
           "#STAR.splice_junctions"          "#STAR.chimeric_junctions" </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get input and output object</span>
t1<span class="op">$</span><span class="kw">get_input</span>(<span class="dt">id =</span> <span class="st">"#winFlankNbins"</span>)</code></pre></div>
<pre><code>type:
- 'null'
- int
label: Flanking regions size
description: =log2(winFlank), where win Flank is the size of the left and right flanking
  regions for each window (int&gt;0).
streamable: no
id: '#winFlankNbins'
inputBinding:
  position: 0
  prefix: --winFlankNbins
  separate: yes
  sbg:cmdInclude: yes
sbg:category: Windows, Anchors, Binning
sbg:toolDefaultValue: '4'
required: no</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">t1<span class="op">$</span><span class="kw">get_input</span>(<span class="dt">name =</span> <span class="st">"ins"</span>)</code></pre></div>
<pre><code>[[1]]
type:
- 'null'
- int
label: Max bins between anchors
description: Max number of bins between two anchors that allows aggregation of anchors
  into one window (int&gt;0).
streamable: no
id: '#winAnchorDistNbins'
inputBinding:
  position: 0
  prefix: --winAnchorDistNbins
  separate: yes
  sbg:cmdInclude: yes
sbg:category: Windows, Anchors, Binning
sbg:toolDefaultValue: '9'
required: no

[[2]]
type:
- 'null'
- int
label: Max insert junctions
description: Maximum number of junction to be inserted to the genome on the fly at
  the mapping stage, including those from annotations and those detected in the 1st
  step of the 2-pass run.
streamable: no
id: '#limitSjdbInsertNsj'
inputBinding:
  position: 0
  prefix: --limitSjdbInsertNsj
  separate: yes
  sbg:cmdInclude: yes
sbg:category: Limits
sbg:toolDefaultValue: '1000000'
required: no</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">t1<span class="op">$</span><span class="kw">get_output</span>(<span class="dt">id =</span> <span class="st">"#aligned_reads"</span>)</code></pre></div>
<pre><code>type:
- 'null'
- File
label: Aligned SAM/BAM
description: Aligned sequence in SAM/BAM format.
streamable: no
id: '#aligned_reads'
outputBinding:
  glob:
    engine: '#cwl-js-engine'
    script: |-
      {
        if ($job.inputs.outSortingType == 'SortedByCoordinate') {
          sort_name = '.sortedByCoord'
        }
        else {
          sort_name = ''
        }
        if ($job.inputs.outSAMtype == 'BAM') {
          sam_name = "*.Aligned".concat( sort_name, '.out.bam')
        }
        else {
          sam_name = "*.Aligned.out.sam"
        }
        return sam_name
      }
    class: Expression
sbg:fileTypes: SAM, BAM</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">t1<span class="op">$</span><span class="kw">get_output</span>(<span class="dt">name =</span> <span class="st">"gene"</span>)</code></pre></div>
<pre><code>type:
- 'null'
- File
label: Reads per gene
description: File with number of reads per gene. A read is counted if it overlaps
  (1nt or more) one and only one gene.
streamable: no
id: '#reads_per_gene'
outputBinding:
  glob: '*ReadsPerGene*'
sbg:fileTypes: TAB</code></pre>
</div>
<div id="create-your-own-tool-in-r" class="section level2">
<h2 class="hasAnchor">
<a href="#create-your-own-tool-in-r" class="anchor"></a>Create your own tool in R</h2>
<div id="introduction" class="section level3">
<h3 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h3>
<p>Before we continue, this is how it looks like for full tool description, you don’t always need to describe all those details, following section will walk you through simple examples to full examples like this one.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fl &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"docker/rnaseqGene/rabix"</span>, <span class="st">"generator.R"</span>, <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
<span class="kw">cat</span>(<span class="kw">readLines</span>(fl), <span class="dt">sep =</span> <span class="st">'</span><span class="ch">\n</span><span class="st">'</span>)</code></pre></div>
<pre><code>library("sevenbridges")

rbx &lt;- Tool(id = "rnaseqGene",
            label = "rnaseqgene",
            description = "A RNA-seq Differiencial Expression Flow and Report",
            hints = requirements(docker(pull = "tengfei/rnaseqgene"), cpu(1), mem(2000)),
            baseCommand = "performDE.R",
            inputs = list(
                input(
                    id = "bamfiles", label = "bam files",
                    description = "a list of bam files",
                    type = "File...",  ## or type = ItemArray("File")
                    prefix = "--bamfiles",
                    required = TRUE,
                    itemSeparator = ","
                ),
                input(
                    id = "design", label = "design matrix",
                    type = "File",
                    required = TRUE,
                    prefix = "--design"
                ),
                input(
                    id = "gtffile", label =  "gene feature files",
                    type = "File",
                    stageInput = "copy",
                    required = TRUE,
                    prefix = "--gtffile"
                ),
                input(
                    id = "format", label =  "report foramt html or pdf",
                    type = enum("format", c("pdf", "html")),
                    prefix = "--format"
                )
            ),
            outputs = list(
                output(id = "report", label = "report",
                       description = "A reproducible report created by Rmarkdown",
                       glob = Expression(engine = "#cwl-js-engine",
                                         script = "x = $job[['inputs']][['format']];
                                         if(x == 'undefined' || x == null){
                                         x = 'html';
                                         };
                                         'rnaseqGene.' +  x")),
                output(id = "heatmap", label = "heatmap",
                       description = "A heatmap plot to show the Euclidean distance between samples",
                       glob = "heatmap.pdf"),
                output(id = "count", label = "count",
                       description = "Reads counts matrix",
                       glob = "count.csv"),
                output(id = "de", label = "Differential expression table",
                       description = "Differential expression table",
                       glob = "de.csv")
                       ))

fl &lt;- "inst/docker/rnaseqGene/rabix/rnaseqGene.json"
write(rbx$toJSON(pretty = TRUE), fl)</code></pre>
<p>Now let’s break it down:</p>
<p>Some key arguments used in <code>Tool</code> function.</p>
<ul>
<li>
<strong>baseCommand</strong>: Specifies the program to execute.</li>
<li>
<strong>stdout</strong>: Capture the command’s standard output stream to a file written to the designated output directory. You don’t need this, if you specify output files to collect.</li>
<li>
<strong>inputs</strong>: inputs for your command line</li>
<li>
<strong>outputs</strong>: outputs you want to collect</li>
<li>
<strong>Requirements</strong> and <strong>hints</strong>: in short, hints are not <em>required</em> for execution. We now accept following requirement items <code>cpu</code>, <code>mem</code>, <code>docker</code>, <code>fileDef</code>; and you can easily construct them via <code><a href="../reference/requirements.html">requirements()</a></code> constructor. This is how you describe the resources you need to execute the tool, so the system knows what type of instances suit your case best.</li>
</ul>
<p>To specify inputs and outpus, usually your command line interface accept extra arguments as input, for example, file(s), string, enum, int, float, boolean. So to specify that in your tool, you can use <code>input</code> function, then pass it to the <code>inputs</code> arguments as a list or single item. You can even construct them as data.frame with less flexibility. <code><a href="../reference/shorthand-cwl.html">input()</a></code> require arguments <code>id</code> and <code>type</code>. <code><a href="../reference/shorthand-cwl.html">output()</a></code> require arguments <code>id</code> because <code>type</code> by default is file.</p>
<p>There are some special type: ItemArray and enum. For ItemArray the type could be an array of single type, the most common case is that if your input is a list of files, you can do something like <code>type = ItemArray("File")</code> or as simple as <code>type = "File..."</code> to diffenciate from a single file input. When you add “…” suffix, R will know it’s an <code>ItemArray</code>.</p>
<p>We also provide an <strong>enum</strong> type, when you specify the enum, please pass the required name and symbols like this <code>type = enum("format", c("pdf", "html"))</code> then in the UI on the platform you will be poped with drop down when you execute the task.</p>
<p>Now let’s work though from simple case to most flexible case.</p>
</div>
<div id="using-existing-docker-images-and-command" class="section level3">
<h3 class="hasAnchor">
<a href="#using-existing-docker-images-and-command" class="anchor"></a>Using existing Docker images and command</h3>
<p>If you already have a Docker image in mind that provide the functionality you need, you can just use it. The <code>baseCommand</code> is the command line you want to execute in that container. <code>stdout</code> specify the output file you want to capture the standard output and collect it on the platform.</p>
<p>In this simple example, I know Docker image <code>rocker/r-base</code> has a function called <code>runif</code> I can directly called in command line with <code>Rscript -e</code>. Then I want the ouput is collected in <code>stdout</code> and ask the file system to capture the files matches “*.txt“. Please pay attention to this, you tool may produce many intermediate files in current folder, if you don’t tell which output you need, they will all be ignored, so make sure you collect those files via <code>outputs</code> parameter.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"sevenbridges"</span>)
rbx &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Tool-class.html">Tool</a></span>(<span class="dt">id =</span> <span class="st">"runif"</span>,
            <span class="dt">label =</span> <span class="st">"runif"</span>,
            <span class="dt">hints =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(<span class="kw"><a href="../reference/requirements.html">docker</a></span>(<span class="dt">pull =</span> <span class="st">"rocker/r-base"</span>)),
            <span class="dt">baseCommand =</span> <span class="st">"Rscript -e 'runif(100)'"</span>,
            <span class="dt">stdout =</span> <span class="st">"output.txt"</span>,
            <span class="dt">outputs =</span> <span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"random"</span>, <span class="dt">glob =</span> <span class="st">"*.txt"</span>))

rbx</code></pre></div>
<pre><code>sbg:id: runif
id: '#runif'
inputs: []
outputs:
- type:
  - 'null'
  - File
  label: ''
  description: ''
  streamable: no
  default: ''
  id: '#random'
  outputBinding:
    glob: '*.txt'
requirements: []
hints:
- class: DockerRequirement
  dockerPull: rocker/r-base
label: runif
class: CommandLineTool
baseCommand:
- Rscript -e 'runif(100)'
arguments: []
stdout: output.txt</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rbx<span class="op">$</span><span class="kw">toJSON</span>()</code></pre></div>
<pre><code>{"sbg:id":"runif","id":"#runif","inputs":[],"outputs":[{"type":["null","File"],"label":"","description":"","streamable":false,"default":"","id":"#random","outputBinding":{"glob":"*.txt"}}],"requirements":[],"hints":[{"class":"DockerRequirement","dockerPull":"rocker/r-base"}],"label":"runif","class":"CommandLineTool","baseCommand":["Rscript -e 'runif(100)'"],"arguments":[],"stdout":"output.txt"} </code></pre>
<p>By default the tool object shows YAML, but you can simply convert it to JSON and copy it to your seven bridges platform graphic editor by importing JSON.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rbx<span class="op">$</span><span class="kw">toJSON</span>()</code></pre></div>
<pre><code>{"sbg:id":"runif","id":"#runif","inputs":[],"outputs":[{"type":["null","File"],"label":"","description":"","streamable":false,"default":"","id":"#random","outputBinding":{"glob":"*.txt"}}],"requirements":[],"hints":[{"class":"DockerRequirement","dockerPull":"rocker/r-base"}],"label":"runif","class":"CommandLineTool","baseCommand":["Rscript -e 'runif(100)'"],"arguments":[],"stdout":"output.txt"} </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rbx<span class="op">$</span><span class="kw">toJSON</span>(<span class="dt">pretty =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>{
  "sbg:id": "runif",
  "id": "#runif",
  "inputs": [],
  "outputs": [
    {
      "type": ["null", "File"],
      "label": "",
      "description": "",
      "streamable": false,
      "default": "",
      "id": "#random",
      "outputBinding": {
        "glob": "*.txt"
      }
    }
  ],
  "requirements": [],
  "hints": [
    {
      "class": "DockerRequirement",
      "dockerPull": "rocker/r-base"
    }
  ],
  "label": "runif",
  "class": "CommandLineTool",
  "baseCommand": [
    "Rscript -e 'runif(100)'"
  ],
  "arguments": [],
  "stdout": "output.txt"
} </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rbx<span class="op">$</span><span class="kw">toYAML</span>()</code></pre></div>
<pre><code>[1] "sbg:id: runif\nid: '#runif'\ninputs: []\noutputs:\n- type:\n  - 'null'\n  - File\n  label: ''\n  description: ''\n  streamable: no\n  default: ''\n  id: '#random'\n  outputBinding:\n    glob: '*.txt'\nrequirements: []\nhints:\n- class: DockerRequirement\n  dockerPull: rocker/r-base\nlabel: runif\nclass: CommandLineTool\nbaseCommand:\n- Rscript -e 'runif(100)'\narguments: []\nstdout: output.txt\n"</code></pre>
</div>
<div id="add-customized-script-to-existing-docker-image" class="section level3">
<h3 class="hasAnchor">
<a href="#add-customized-script-to-existing-docker-image" class="anchor"></a>Add customized script to existing Docker image</h3>
<p>Now you make want to run your own R script, but you still don’t want to create new command line and a new Docker image. You just want to run your script with new input files in existing container, it’s time to introduce <code>fileDef</code>. You can either directly write script as string or just import a R file to <code>content</code>. And provided as <code>requirements</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Make a new file</span>
fd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/requirements.html">fileDef</a></span>(<span class="dt">name =</span> <span class="st">"runif.R"</span>,
              <span class="dt">content =</span> <span class="st">"set.seed(1)</span>
<span class="st">                   runif(100)"</span>)

<span class="co"># read via reader</span>
.srcfile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"docker/sevenbridges/src/runif.R"</span>, <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
<span class="kw">library</span>(<span class="st">"readr"</span>)
fd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/requirements.html">fileDef</a></span>(<span class="dt">name =</span> <span class="st">"runif.R"</span>,
              <span class="dt">content =</span> <span class="kw"><a href="http://www.rdocumentation.org/packages/readr/topics/read_file">read_file</a></span>(.srcfile))

<span class="co"># add script to your tool</span>
rbx &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Tool-class.html">Tool</a></span>(<span class="dt">id =</span> <span class="st">"runif"</span>,
            <span class="dt">label =</span> <span class="st">"runif"</span>,
            <span class="dt">hints =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(<span class="kw"><a href="../reference/requirements.html">docker</a></span>(<span class="dt">pull =</span> <span class="st">"rocker/r-base"</span>)),
            <span class="dt">requirements =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(fd),
            <span class="dt">baseCommand =</span> <span class="st">"Rscript runif.R"</span>,
            <span class="dt">stdout =</span> <span class="st">"output.txt"</span>,
            <span class="dt">outputs =</span> <span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"random"</span>, <span class="dt">glob =</span> <span class="st">"*.txt"</span>))</code></pre></div>
<p>How about multiple script?</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># or simply readLines</span>
.srcfile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"docker/sevenbridges/src/runif.R"</span>, <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
<span class="kw">library</span>(<span class="st">"readr"</span>)
fd1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/requirements.html">fileDef</a></span>(<span class="dt">name =</span> <span class="st">"runif.R"</span>,
              <span class="dt">content =</span> <span class="kw"><a href="http://www.rdocumentation.org/packages/readr/topics/read_file">read_file</a></span>(.srcfile))
fd2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/requirements.html">fileDef</a></span>(<span class="dt">name =</span> <span class="st">"runif2.R"</span>,
              <span class="dt">content =</span> <span class="st">"set.seed(1)</span>
<span class="st">                   runif(100)"</span>)

rbx &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Tool-class.html">Tool</a></span>(<span class="dt">id =</span> <span class="st">"runif_twoscript"</span>,
            <span class="dt">label =</span> <span class="st">"runif_twoscript"</span>,
            <span class="dt">hints =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(<span class="kw"><a href="../reference/requirements.html">docker</a></span>(<span class="dt">pull =</span> <span class="st">"rocker/r-base"</span>)),
            <span class="dt">requirements =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(fd1, fd2),
            <span class="dt">baseCommand =</span> <span class="st">"Rscript runif.R"</span>,
            <span class="dt">stdout =</span> <span class="st">"output.txt"</span>,
            <span class="dt">outputs =</span> <span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"random"</span>, <span class="dt">glob =</span> <span class="st">"*.txt"</span>))</code></pre></div>
</div>
<div id="create-formal-interface-for-your-command-line" class="section level3">
<h3 class="hasAnchor">
<a href="#create-formal-interface-for-your-command-line" class="anchor"></a>Create formal interface for your command line</h3>
<p>All those examples above, many parameters are hard-coded in your script, you don’t have flexiblity to control how many numbers to generate. Most often, your tools or command line tools expose some inputs arguments to users. You need a better way to describe a command line with input/output.</p>
<p>Now we bring the example to next level, for example, I prepare a Docker image called “tengfei/runif” on Docker Hub, this container has a exeutable command called “runif.R”, you don’t have to know what’s inside, you only have to know when you run the command line in that container it looks like this</p>
<pre><code>runif.R --n=100 --max=100 --min=1 --seed=123</code></pre>
<p>This command outpus two files directly, so you don’t need standard output to capture random number.</p>
<ul>
<li>output.txt</li>
<li>report.html</li>
</ul>
<p>So the goal here is to describe this command and expose all input parameters and collect all two files.</p>
<p>To define input, you can specify</p>
<ul>
<li>
<strong>id</strong> : unique identifier to this input node.</li>
<li>
<strong>description</strong>: description, also visible on UI.</li>
<li>
<strong>type</strong>: required to specify input types, files, integer, or character.</li>
<li>
<strong>label</strong>: human readable label for this input node.</li>
<li>
<strong>prefix</strong>: the prefix in command line for this input parameter.</li>
<li>
<strong>default</strong>: default value for this input.</li>
<li>
<strong>required</strong>: is this input parameter required or not. If required, when you execte the tool you have to provide a value for the parameter.</li>
<li>
<strong>cmdInclude</strong>: included in command line or not.</li>
</ul>
<p>Output is similar, espeicaly when you want to collect file, you can use <code>glob</code> for pattern matching.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># pass a input list</span>
in.lst &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="kw"><a href="../reference/shorthand-cwl.html">input</a></span>(<span class="dt">id =</span> <span class="st">"number"</span>,
                     <span class="dt">description =</span> <span class="st">"number of observations"</span>,
                     <span class="dt">type =</span> <span class="st">"integer"</span>,
                     <span class="dt">label =</span> <span class="st">"number"</span>,
                     <span class="dt">prefix =</span> <span class="st">"--n"</span>,
                     <span class="dt">default =</span> <span class="dv">1</span>,
                     <span class="dt">required =</span> <span class="ot">TRUE</span>,
                     <span class="dt">cmdInclude =</span> <span class="ot">TRUE</span>),
               <span class="kw"><a href="../reference/shorthand-cwl.html">input</a></span>(<span class="dt">id =</span> <span class="st">"min"</span>,
                     <span class="dt">description =</span> <span class="st">"lower limits of the distribution"</span>,
                     <span class="dt">type =</span> <span class="st">"float"</span>,
                     <span class="dt">label =</span> <span class="st">"min"</span>,
                     <span class="dt">prefix =</span> <span class="st">"--min"</span>,
                     <span class="dt">default =</span> <span class="dv">0</span>),
               <span class="kw"><a href="../reference/shorthand-cwl.html">input</a></span>(<span class="dt">id =</span> <span class="st">"max"</span>,
                     <span class="dt">description =</span> <span class="st">"upper limits of the distribution"</span>,
                     <span class="dt">type =</span> <span class="st">"float"</span>,
                     <span class="dt">label =</span> <span class="st">"max"</span>,
                     <span class="dt">prefix =</span> <span class="st">"--max"</span>,
                     <span class="dt">default =</span> <span class="dv">1</span>),
               <span class="kw"><a href="../reference/shorthand-cwl.html">input</a></span>(<span class="dt">id =</span> <span class="st">"seed"</span>,
                     <span class="dt">description =</span> <span class="st">"seed with set.seed"</span>,
                     <span class="dt">type =</span> <span class="st">"float"</span>,
                     <span class="dt">label =</span> <span class="st">"seed"</span>,
                     <span class="dt">prefix =</span> <span class="st">"--seed"</span>,
                     <span class="dt">default =</span> <span class="dv">1</span>))

<span class="co"># the same method for outputs</span>
out.lst &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"random"</span>,
                       <span class="dt">type =</span> <span class="st">"file"</span>,
                       <span class="dt">label =</span> <span class="st">"output"</span>,
                       <span class="dt">description =</span> <span class="st">"random number file"</span>,
                       <span class="dt">glob =</span> <span class="st">"*.txt"</span>),
                <span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"report"</span>,
                       <span class="dt">type =</span> <span class="st">"file"</span>,
                       <span class="dt">label =</span> <span class="st">"report"</span>,
                       <span class="dt">glob =</span> <span class="st">"*.html"</span>))

rbx &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Tool-class.html">Tool</a></span>(<span class="dt">id =</span> <span class="st">"runif"</span>,
            <span class="dt">label =</span> <span class="st">"Random number generator"</span>,
            <span class="dt">hints =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(<span class="kw"><a href="../reference/requirements.html">docker</a></span>(<span class="dt">pull =</span> <span class="st">"tengfei/runif"</span>)),
            <span class="dt">baseCommand =</span> <span class="st">"runif.R"</span>,
            <span class="dt">inputs =</span> in.lst,  <span class="co"># or ins.df</span>
            <span class="dt">outputs =</span> out.lst)</code></pre></div>
<p>Alternatively you can use data.frame as example for input and output, but it’s less flexible.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">in.df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">id =</span> <span class="kw">c</span>(<span class="st">"number"</span>, <span class="st">"min"</span>, <span class="st">"max"</span>, <span class="st">"seed"</span>),
                    <span class="dt">description =</span> <span class="kw">c</span>(<span class="st">"number of observation"</span>,
                                    <span class="st">"lower limits of the distribution"</span>,
                                    <span class="st">"upper limits of the distribution"</span>,
                                    <span class="st">"seed with set.seed"</span>),
                    <span class="dt">type =</span> <span class="kw">c</span>(<span class="st">"integer"</span>, <span class="st">"float"</span>, <span class="st">"float"</span>, <span class="st">"float"</span>),
                    <span class="dt">label =</span> <span class="kw">c</span>(<span class="st">"number"</span> ,<span class="st">"min"</span>, <span class="st">"max"</span>, <span class="st">"seed"</span>),
                    <span class="dt">prefix =</span> <span class="kw">c</span>(<span class="st">"--n"</span>, <span class="st">"--min"</span>, <span class="st">"--max"</span>, <span class="st">"--seed"</span>),
                    <span class="dt">default =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">10</span>, <span class="dv">123</span>),
                    <span class="dt">required =</span> <span class="kw">c</span>(<span class="ot">TRUE</span>, <span class="ot">FALSE</span>, <span class="ot">FALSE</span>, <span class="ot">FALSE</span>))

out.df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">id =</span> <span class="kw">c</span>(<span class="st">"random"</span>, <span class="st">"report"</span>),
                     <span class="dt">type =</span> <span class="kw">c</span>(<span class="st">"file"</span>, <span class="st">"file"</span>),
                     <span class="dt">glob =</span> <span class="kw">c</span>(<span class="st">"*.txt"</span>, <span class="st">"*.html"</span>))

rbx &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Tool-class.html">Tool</a></span>(<span class="dt">id =</span> <span class="st">"runif"</span>,
            <span class="dt">label =</span> <span class="st">"Random number generator"</span>,
            <span class="dt">hints =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(<span class="kw"><a href="../reference/requirements.html">docker</a></span>(<span class="dt">pull =</span> <span class="st">"tengfei/runif"</span>),
                                 <span class="kw"><a href="../reference/requirements.html">cpu</a></span>(<span class="dv">1</span>), <span class="kw"><a href="../reference/requirements.html">mem</a></span>(<span class="dv">2000</span>)),
            <span class="dt">baseCommand =</span> <span class="st">"runif.R"</span>,
            <span class="dt">inputs =</span> in.df,  <span class="co"># or ins.df</span>
            <span class="dt">outputs =</span> out.df)</code></pre></div>
</div>
<div id="quick-command-line-interface-with-commandargs-position-and-named-args" class="section level3">
<h3 class="hasAnchor">
<a href="#quick-command-line-interface-with-commandargs-position-and-named-args" class="anchor"></a>Quick command line interface with <code>commandArgs</code> (position and named args)</h3>
<p>Now you must be wondering, I have a Docker container with R, but I don’t have any existing command line that I could directly use. Can I provide a script with a formal and quick command line interface to make an App for existing container. The anwser is yes. When you add script to your tool, you can always use some trick to do so, one popular one you may already head of is <code>commandArgs</code>. More formal one is called “docopt” which I will show you later.</p>
<p>Suppose you have a R script “runif2spin.R” with three arguments using position mapping</p>
<ol style="list-style-type: decimal">
<li><code>numbers</code></li>
<li><code>min</code></li>
<li><code>max</code></li>
</ol>
<p>My base command will be somethine like</p>
<pre><code>Rscript runif2spin.R 10 30 50</code></pre>
<p>This is how you do in your R script</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fl &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"docker/sevenbridges/src"</span>, <span class="st">"runif2spin.R"</span>,
                  <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
<span class="kw">cat</span>(<span class="kw">readLines</span>(fl), <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</code></pre></div>
<pre><code>#' ---
#' title: "Uniform random number generator example"
#' output:
#'     html_document:
#'     toc: true
#' number_sections: true
#' ---

#' # summary report
#'
#' This is a random number generator

#+
args = commandArgs(TRUE)

r = runif(n   = as.integer(args[1]),
          min = as.numeric(args[2]),
          max = as.numeric(args[3]))
head(r)
summary(r)
hist(r)</code></pre>
<p>Ignore the comment part, I will introduce spin/stich later.</p>
<p>Then just describe my tool in this way, add your script as you learned in previous sections.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"readr"</span>)
fd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/requirements.html">fileDef</a></span>(<span class="dt">name =</span> <span class="st">"runif.R"</span>,
              <span class="dt">content =</span> <span class="kw"><a href="http://www.rdocumentation.org/packages/readr/topics/read_file">read_file</a></span>(fl))

rbx &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Tool-class.html">Tool</a></span>(<span class="dt">id =</span> <span class="st">"runif"</span>,
            <span class="dt">label =</span> <span class="st">"runif"</span>,
            <span class="dt">hints =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(<span class="kw"><a href="../reference/requirements.html">docker</a></span>(<span class="dt">pull =</span> <span class="st">"rocker/r-base"</span>),
                                 <span class="kw"><a href="../reference/requirements.html">cpu</a></span>(<span class="dv">1</span>), <span class="kw"><a href="../reference/requirements.html">mem</a></span>(<span class="dv">2000</span>)),
            <span class="dt">requirements =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(fd),
            <span class="dt">baseCommand =</span> <span class="st">"Rscript runif.R"</span>,
            <span class="dt">stdout =</span> <span class="st">"output.txt"</span>,
            <span class="dt">inputs =</span> <span class="kw">list</span>(<span class="kw"><a href="../reference/shorthand-cwl.html">input</a></span>(<span class="dt">id       =</span> <span class="st">"number"</span>,
                                <span class="dt">type     =</span> <span class="st">"integer"</span>,
                                <span class="dt">position =</span> <span class="dv">1</span>),
                          <span class="kw"><a href="../reference/shorthand-cwl.html">input</a></span>(<span class="dt">id       =</span> <span class="st">"min"</span>,
                                <span class="dt">type     =</span> <span class="st">"float"</span>,
                                <span class="dt">position =</span> <span class="dv">2</span>),
                          <span class="kw"><a href="../reference/shorthand-cwl.html">input</a></span>(<span class="dt">id       =</span> <span class="st">"max"</span>,
                                <span class="dt">type     =</span> <span class="st">"float"</span>,
                                <span class="dt">position =</span> <span class="dv">3</span>)),
            <span class="dt">outputs =</span> <span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"random"</span>, <span class="dt">glob =</span> <span class="st">"output.txt"</span>))</code></pre></div>
<p>How about named argumentments? I will still recommend use “docopt” package, but for simple way. You want command line looks like this</p>
<pre><code>Rscript runif_args.R --n=10 --min=30 --max=50</code></pre>
<p>Here is how you do in R script.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fl &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"docker/sevenbridges/src"</span>, <span class="st">"runif_args.R"</span>, <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
<span class="kw">cat</span>(<span class="kw">readLines</span>(fl), <span class="dt">sep =</span> <span class="st">'</span><span class="ch">\n</span><span class="st">'</span>)</code></pre></div>
<pre><code>#' ---
#' title: "Uniform random number generator example"
#' output:
#'     html_document:
#'     toc: true
#' number_sections: true
#' ---

#' # summary report
#'
#' This is a random number generator

#+
args &lt;- commandArgs(TRUE)

## quick hack to split named arguments
splitArgs &lt;- function(x) {
    res &lt;- do.call(rbind, lapply(x, function(i){
        res &lt;- strsplit(i, "=")[[1]]
        nm &lt;- gsub("-+", "",res[1])
        c(nm, res[2])
    }))
    .r &lt;- res[,2]
    names(.r) &lt;- res[,1]
    .r
}
args &lt;- splitArgs(args)

#+
r &lt;- runif(n   = as.integer(args["n"]),
           min = as.numeric(args["min"]),
           max = as.numeric(args["max"]))
summary(r)
hist(r)
write.csv(r, file = "out.csv")</code></pre>
<p>Then just describe my tool in this way, note, I use <code>separate=FALSE</code> and add <code>=</code> to my prefix as a hack.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"readr"</span>)
fd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/requirements.html">fileDef</a></span>(<span class="dt">name =</span> <span class="st">"runif.R"</span>,
              <span class="dt">content =</span> <span class="kw"><a href="http://www.rdocumentation.org/packages/readr/topics/read_file">read_file</a></span>(fl))

rbx &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Tool-class.html">Tool</a></span>(<span class="dt">id =</span> <span class="st">"runif"</span>,
            <span class="dt">label =</span> <span class="st">"runif"</span>,
            <span class="dt">hints =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(<span class="kw"><a href="../reference/requirements.html">docker</a></span>(<span class="dt">pull =</span> <span class="st">"rocker/r-base"</span>),
                                 <span class="kw"><a href="../reference/requirements.html">cpu</a></span>(<span class="dv">1</span>), <span class="kw"><a href="../reference/requirements.html">mem</a></span>(<span class="dv">2000</span>)),
            <span class="dt">requirements =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(fd),
            <span class="dt">baseCommand =</span> <span class="st">"Rscript runif.R"</span>,
            <span class="dt">stdout =</span> <span class="st">"output.txt"</span>,
            <span class="dt">inputs =</span> <span class="kw">list</span>(<span class="kw"><a href="../reference/shorthand-cwl.html">input</a></span>(<span class="dt">id       =</span> <span class="st">"number"</span>,
                                <span class="dt">type     =</span> <span class="st">"integer"</span>,
                                <span class="dt">separate =</span> <span class="ot">FALSE</span>,
                                <span class="dt">prefix   =</span> <span class="st">"--n="</span>),
                          <span class="kw"><a href="../reference/shorthand-cwl.html">input</a></span>(<span class="dt">id       =</span> <span class="st">"min"</span>,
                                <span class="dt">type     =</span> <span class="st">"float"</span>,
                                <span class="dt">separate =</span> <span class="ot">FALSE</span>,
                                <span class="dt">prefix   =</span> <span class="st">"--min="</span>),
                          <span class="kw"><a href="../reference/shorthand-cwl.html">input</a></span>(<span class="dt">id       =</span> <span class="st">"max"</span>,
                                <span class="dt">type     =</span> <span class="st">"float"</span>,
                                <span class="dt">separate =</span> <span class="ot">FALSE</span>,
                                <span class="dt">prefix   =</span> <span class="st">"--max="</span>)),
            <span class="dt">outputs =</span> <span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"random"</span>, <span class="dt">glob =</span> <span class="st">"output.txt"</span>))</code></pre></div>
</div>
<div id="docopt-a-better-and-formal-way-to-make-command-line-interface" class="section level3">
<h3 class="hasAnchor">
<a href="#docopt-a-better-and-formal-way-to-make-command-line-interface" class="anchor"></a>docopt: a better and formal way to make command line interface</h3>
</div>
<div id="generate-reports" class="section level3">
<h3 class="hasAnchor">
<a href="#generate-reports" class="anchor"></a>Generate reports</h3>
<p><strong>Quick report: Spin and Stich</strong></p>
<p>You can use spin/stich from knitr to generate report directly from a Rscript with special format. For example, let’s use above example</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fl &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"docker/sevenbridges/src"</span>, <span class="st">"runif_args.R"</span>, <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
<span class="kw">cat</span>(<span class="kw">readLines</span>(fl), <span class="dt">sep =</span> <span class="st">'</span><span class="ch">\n</span><span class="st">'</span>)</code></pre></div>
<pre><code>#' ---
#' title: "Uniform random number generator example"
#' output:
#'     html_document:
#'     toc: true
#' number_sections: true
#' ---

#' # summary report
#'
#' This is a random number generator

#+
args &lt;- commandArgs(TRUE)

## quick hack to split named arguments
splitArgs &lt;- function(x) {
    res &lt;- do.call(rbind, lapply(x, function(i){
        res &lt;- strsplit(i, "=")[[1]]
        nm &lt;- gsub("-+", "",res[1])
        c(nm, res[2])
    }))
    .r &lt;- res[,2]
    names(.r) &lt;- res[,1]
    .r
}
args &lt;- splitArgs(args)

#+
r &lt;- runif(n   = as.integer(args["n"]),
           min = as.numeric(args["min"]),
           max = as.numeric(args["max"]))
summary(r)
hist(r)
write.csv(r, file = "out.csv")</code></pre>
<p>You command is something like this</p>
<pre><code>Rscript -e "rmarkdown::render(knitr::spin('runif_args.R', FALSE))" --args --n=100 --min=30 --max=50</code></pre>
<p>And so I describe my tool like this with Docker image <code>rocker/hadleyverse</code> this contians knitr and rmarkdown package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"readr"</span>)
fd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/requirements.html">fileDef</a></span>(<span class="dt">name =</span> <span class="st">"runif.R"</span>,
              <span class="dt">content =</span> <span class="kw"><a href="http://www.rdocumentation.org/packages/readr/topics/read_file">read_file</a></span>(fl))

rbx &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Tool-class.html">Tool</a></span>(<span class="dt">id =</span> <span class="st">"runif"</span>,
            <span class="dt">label =</span> <span class="st">"runif"</span>,
            <span class="dt">hints =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(<span class="kw"><a href="../reference/requirements.html">docker</a></span>(<span class="dt">pull =</span> <span class="st">"rocker/hadleyverse"</span>),
                                 <span class="kw"><a href="../reference/requirements.html">cpu</a></span>(<span class="dv">1</span>), <span class="kw"><a href="../reference/requirements.html">mem</a></span>(<span class="dv">2000</span>)),
            <span class="dt">requirements =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(fd),
            <span class="dt">baseCommand =</span> <span class="st">"Rscript -e </span><span class="ch">\"</span><span class="st">rmarkdown::render(knitr::spin('runif.R', FALSE))</span><span class="ch">\"</span><span class="st"> --args"</span>,
            <span class="dt">stdout =</span> <span class="st">"output.txt"</span>,
            <span class="dt">inputs =</span> <span class="kw">list</span>(<span class="kw"><a href="../reference/shorthand-cwl.html">input</a></span>(<span class="dt">id       =</span> <span class="st">"number"</span>,
                                <span class="dt">type     =</span> <span class="st">"integer"</span>,
                                <span class="dt">separate =</span> <span class="ot">FALSE</span>,
                                <span class="dt">prefix   =</span> <span class="st">"--n="</span>),
                          <span class="kw"><a href="../reference/shorthand-cwl.html">input</a></span>(<span class="dt">id       =</span> <span class="st">"min"</span>,
                                <span class="dt">type     =</span> <span class="st">"float"</span>,
                                <span class="dt">separate =</span> <span class="ot">FALSE</span>,
                                <span class="dt">prefix   =</span> <span class="st">"--min="</span>),
                          <span class="kw"><a href="../reference/shorthand-cwl.html">input</a></span>(<span class="dt">id       =</span> <span class="st">"max"</span>,
                                <span class="dt">type     =</span> <span class="st">"float"</span>,
                                <span class="dt">separate =</span> <span class="ot">FALSE</span>,
                                <span class="dt">prefix   =</span> <span class="st">"--max="</span>)),
            <span class="dt">outputs =</span> <span class="kw">list</span>(<span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"stdout"</span>, <span class="dt">type =</span> <span class="st">"file"</span>, <span class="dt">glob =</span> <span class="st">"output.txt"</span>),
                           <span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"random"</span>, <span class="dt">type =</span> <span class="st">"file"</span>, <span class="dt">glob =</span> <span class="st">"*.csv"</span>),
                           <span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"report"</span>, <span class="dt">type =</span> <span class="st">"file"</span>, <span class="dt">glob =</span> <span class="st">"*.html"</span>)))</code></pre></div>
<p>You will get a report in the end.</p>
</div>
<div id="misc" class="section level3">
<h3 class="hasAnchor">
<a href="#misc" class="anchor"></a>Misc</h3>
<p><strong>Inherit metadata and additional metadata</strong></p>
<p>Sometimes if you want your output files inherit from particular input file, just use <code>inheritMetadataFrom</code> in your output() call and pass the input file id. If you want to add additional metadata, you could pass <code>metadata</code> a list in your output() function call. For example, I want my output report inherit all metadata from my “bam_file” input node (which I don’t have in this example though) with two additional metadata fields.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">out.lst &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"random"</span>,
                       <span class="dt">type =</span> <span class="st">"file"</span>,
                       <span class="dt">label =</span> <span class="st">"output"</span>,
                       <span class="dt">description =</span> <span class="st">"random number file"</span>,
                       <span class="dt">glob =</span> <span class="st">"*.txt"</span>),
                <span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"report"</span>,
                       <span class="dt">type =</span> <span class="st">"file"</span>,
                       <span class="dt">label =</span> <span class="st">"report"</span>,
                       <span class="dt">glob =</span> <span class="st">"*.html"</span>,
                       <span class="dt">inheritMetadataFrom =</span> <span class="st">"bam_file"</span>,
                       <span class="dt">metadata =</span> <span class="kw">list</span>(<span class="dt">author =</span> <span class="st">"tengfei"</span>,
                                       <span class="dt">sample =</span> <span class="st">"random"</span>)))
out.lst</code></pre></div>
<pre><code>[[1]]
type:
- 'null'
- File
label: output
description: random number file
streamable: no
default: ''
id: '#random'
outputBinding:
  glob: '*.txt'


[[2]]
type:
- 'null'
- File
label: report
description: ''
streamable: no
default: ''
id: '#report'
outputBinding:
  glob: '*.html'
  sbg:inheritMetadataFrom: '#bam_file'
  sbg:metadata:
    author: tengfei
    sample: random</code></pre>
<p><strong>Example with file/files as input node</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fl &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"docker/rnaseqGene/rabix"</span>, <span class="st">"generator.R"</span>, <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
<span class="kw">cat</span>(<span class="kw">readLines</span>(fl), <span class="dt">sep =</span> <span class="st">'</span><span class="ch">\n</span><span class="st">'</span>)</code></pre></div>
<pre><code>library("sevenbridges")

rbx &lt;- Tool(id = "rnaseqGene",
            label = "rnaseqgene",
            description = "A RNA-seq Differiencial Expression Flow and Report",
            hints = requirements(docker(pull = "tengfei/rnaseqgene"), cpu(1), mem(2000)),
            baseCommand = "performDE.R",
            inputs = list(
                input(
                    id = "bamfiles", label = "bam files",
                    description = "a list of bam files",
                    type = "File...",  ## or type = ItemArray("File")
                    prefix = "--bamfiles",
                    required = TRUE,
                    itemSeparator = ","
                ),
                input(
                    id = "design", label = "design matrix",
                    type = "File",
                    required = TRUE,
                    prefix = "--design"
                ),
                input(
                    id = "gtffile", label =  "gene feature files",
                    type = "File",
                    stageInput = "copy",
                    required = TRUE,
                    prefix = "--gtffile"
                ),
                input(
                    id = "format", label =  "report foramt html or pdf",
                    type = enum("format", c("pdf", "html")),
                    prefix = "--format"
                )
            ),
            outputs = list(
                output(id = "report", label = "report",
                       description = "A reproducible report created by Rmarkdown",
                       glob = Expression(engine = "#cwl-js-engine",
                                         script = "x = $job[['inputs']][['format']];
                                         if(x == 'undefined' || x == null){
                                         x = 'html';
                                         };
                                         'rnaseqGene.' +  x")),
                output(id = "heatmap", label = "heatmap",
                       description = "A heatmap plot to show the Euclidean distance between samples",
                       glob = "heatmap.pdf"),
                output(id = "count", label = "count",
                       description = "Reads counts matrix",
                       glob = "count.csv"),
                output(id = "de", label = "Differential expression table",
                       description = "Differential expression table",
                       glob = "de.csv")
                       ))

fl &lt;- "inst/docker/rnaseqGene/rabix/rnaseqGene.json"
write(rbx$toJSON(pretty = TRUE), fl)</code></pre>
<p>Note the stageInput example in the above script, you can set it to “copy” or “link”.</p>
<p><strong>Input node batch mode</strong></p>
<p>Batch by File (the long output has been omitted here):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">f1 =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/app"</span>, <span class="st">"flow_star.json"</span>, <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
f1 =<span class="st"> </span><span class="kw"><a href="../reference/convert_app.html">convert_app</a></span>(f1)
f1<span class="op">$</span><span class="kw">set_batch</span>(<span class="st">"sjdbGTFfile"</span>, <span class="dt">type =</span> <span class="st">"ITEM"</span>)</code></pre></div>
<p>Batch by other critieria such as metadta, following example, is using <code>sample_id</code> and <code>library_id</code> (the long output has been omitted here):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">f1 =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/app"</span>, <span class="st">"flow_star.json"</span>, <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
f1 =<span class="st"> </span><span class="kw"><a href="../reference/convert_app.html">convert_app</a></span>(f1)
f1<span class="op">$</span><span class="kw">set_batch</span>(<span class="st">"sjdbGTFfile"</span>, <span class="kw">c</span>(<span class="st">"metadata.sample_id"</span>, <span class="st">"metadata.library_id"</span>))</code></pre></div>
<pre><code>criteria provided, convert type from ITEM to CRITERIA</code></pre>
<p>When you push your app to the platform, you will see the batch available at task page or workflow editor.</p>
</div>
</div>
</div>
<div id="describe-wokrflow-in-r" class="section level1">
<h1 class="hasAnchor">
<a href="#describe-wokrflow-in-r" class="anchor"></a>Describe Wokrflow in R</h1>
<p><strong><em>Note:</em></strong> The <a href="https://docs.sevenbridges.com/docs/the-tool-editor">GUI Tool Editor</a> on Seven Bridges Platform is more conventient for this purpose.</p>
<div id="import-from-a-json-file" class="section level2">
<h2 class="hasAnchor">
<a href="#import-from-a-json-file" class="anchor"></a>Import from a JSON file</h2>
<p>Yes, you could use the same function <code>convert_app</code> to import JSON files.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">f1 =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/app"</span>, <span class="st">"flow_star.json"</span>, <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
f1 =<span class="st"> </span><span class="kw"><a href="../reference/convert_app.html">convert_app</a></span>(f1)
<span class="co"># show it</span>
<span class="co"># f1</span></code></pre></div>
</div>
<div id="utilities-for-flow-objects" class="section level2">
<h2 class="hasAnchor">
<a href="#utilities-for-flow-objects" class="anchor"></a>Utilities for <code>Flow</code> objects</h2>
<p>Just like <code>Tool</code> object, you also have convenient utils for it, especially useful when you execute task.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">f1 =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/app"</span>, <span class="st">"flow_star.json"</span>, <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
f1 =<span class="st"> </span><span class="kw"><a href="../reference/convert_app.html">convert_app</a></span>(f1)
<span class="co"># input matrix</span>
<span class="kw">head</span>(f1<span class="op">$</span><span class="kw"><a href="../reference/input_output_matrix.html">input_matrix</a></span>())</code></pre></div>
<pre><code>                               id               label    type required
1                    #sjdbGTFfile         sjdbGTFfile File...    FALSE
2                          #fastq               fastq File...     TRUE
3               #genomeFastaFiles    genomeFastaFiles    File     TRUE
4 #sjdbGTFtagExonParentTranscript Exons' parents name  string    FALSE
5       #sjdbGTFtagExonParentGene           Gene name  string    FALSE
6          #winAnchorMultimapNmax    Max loci anchors     int    FALSE
  fileTypes
1      null
2      null
3      null
4      null
5      null
6      null</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># by name</span>
<span class="kw">head</span>(f1<span class="op">$</span><span class="kw"><a href="../reference/input_output_matrix.html">input_matrix</a></span>(<span class="kw">c</span>(<span class="st">"id"</span>, <span class="st">"type"</span>, <span class="st">"required"</span>)))</code></pre></div>
<pre><code>                               id    type required
1                    #sjdbGTFfile File...    FALSE
2                          #fastq File...     TRUE
3               #genomeFastaFiles    File     TRUE
4 #sjdbGTFtagExonParentTranscript  string    FALSE
5       #sjdbGTFtagExonParentGene  string    FALSE
6          #winAnchorMultimapNmax     int    FALSE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># return only required</span>
<span class="kw">head</span>(f1<span class="op">$</span><span class="kw"><a href="../reference/input_output_matrix.html">input_matrix</a></span>(<span class="dt">required =</span> <span class="ot">TRUE</span>))</code></pre></div>
<pre><code>                 id            label    type required fileTypes
2            #fastq            fastq File...     TRUE      null
3 #genomeFastaFiles genomeFastaFiles    File     TRUE      null</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># return everything</span>
<span class="kw">head</span>(f1<span class="op">$</span><span class="kw"><a href="../reference/input_output_matrix.html">input_matrix</a></span>(<span class="ot">NULL</span>))</code></pre></div>
<pre><code>                               id    type required fileTypes
1                    #sjdbGTFfile File...    FALSE      null
2                          #fastq File...     TRUE      null
3               #genomeFastaFiles    File     TRUE      null
4 #sjdbGTFtagExonParentTranscript  string    FALSE      null
5       #sjdbGTFtagExonParentGene  string    FALSE      null
6          #winAnchorMultimapNmax     int    FALSE      null
                label                       category stageInput streamable
1         sjdbGTFfile                           null       null      FALSE
2               fastq                           null       null      FALSE
3    genomeFastaFiles                           null       null      FALSE
4 Exons' parents name Splice junctions db parameters       null      FALSE
5           Gene name Splice junctions db parameters       null      FALSE
6    Max loci anchors      Windows, Anchors, Binning       null      FALSE
   sbg.x    sbg.y sbg.includeInPorts
1 160.50 195.0833                 NA
2 164.25 323.7500               TRUE
3 167.75 469.9999                 NA
4 200.00 350.0000                 NA
5 200.00 400.0000                 NA
6 200.00 450.0000                 NA
                                                description
1                                                      &lt;NA&gt;
2                                                      &lt;NA&gt;
3                                                      &lt;NA&gt;
4         Tag name to be used as exons' transcript-parents.
5               Tag name to be used as exons' gene-parents.
6 Max number of loci anchors are allowed to map to (int&gt;0).
  sbg.toolDefaultValue
1                 &lt;NA&gt;
2                 &lt;NA&gt;
3                 &lt;NA&gt;
4        transcript_id
5              gene_id
6                   50
                                                link_to
1 #STAR_Genome_Generate.sjdbGTFfile | #STAR.sjdbGTFfile
2                     #SBG_FASTQ_Quality_Detector.fastq
3                #STAR_Genome_Generate.genomeFastaFiles
4  #STAR_Genome_Generate.sjdbGTFtagExonParentTranscript
5        #STAR_Genome_Generate.sjdbGTFtagExonParentGene
6                           #STAR.winAnchorMultimapNmax</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># return a output matrix with more informtion</span>
<span class="kw">head</span>(f1<span class="op">$</span><span class="kw"><a href="../reference/input_output_matrix.html">output_matrix</a></span>())</code></pre></div>
<pre><code>                            id                       label    type
1              #unmapped_reads              unmapped_reads File...
2 #transcriptome_aligned_reads transcriptome_aligned_reads    File
3            #splice_junctions            splice_junctions    File
4              #reads_per_gene              reads_per_gene    File
5                   #log_files                   log_files File...
6          #chimeric_junctions          chimeric_junctions    File
  fileTypes
1      null
2      null
3      null
4      null
5      null
6      null</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># return only a few fields</span>
<span class="kw">head</span>(f1<span class="op">$</span><span class="kw"><a href="../reference/input_output_matrix.html">output_matrix</a></span>(<span class="kw">c</span>(<span class="st">"id"</span>, <span class="st">"type"</span>)))</code></pre></div>
<pre><code>                            id    type
1              #unmapped_reads File...
2 #transcriptome_aligned_reads    File
3            #splice_junctions    File
4              #reads_per_gene    File
5                   #log_files File...
6          #chimeric_junctions    File</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># return everything</span>
<span class="kw">head</span>(f1<span class="op">$</span><span class="kw"><a href="../reference/input_output_matrix.html">output_matrix</a></span>(<span class="ot">NULL</span>))</code></pre></div>
<pre><code>                            id                       label    type
1              #unmapped_reads              unmapped_reads File...
2 #transcriptome_aligned_reads transcriptome_aligned_reads    File
3            #splice_junctions            splice_junctions    File
4              #reads_per_gene              reads_per_gene    File
5                   #log_files                   log_files File...
6          #chimeric_junctions          chimeric_junctions    File
  fileTypes required                            source streamable
1      null    FALSE              #STAR.unmapped_reads      FALSE
2      null    FALSE #STAR.transcriptome_aligned_reads      FALSE
3      null    FALSE            #STAR.splice_junctions      FALSE
4      null    FALSE              #STAR.reads_per_gene      FALSE
5      null    FALSE                   #STAR.log_files      FALSE
6      null    FALSE          #STAR.chimeric_junctions      FALSE
  sbg.includeInPorts     sbg.x     sbg.y                           link_to
1               TRUE  766.2498 159.58331              #STAR.unmapped_reads
2               TRUE 1118.9998  86.58332 #STAR.transcriptome_aligned_reads
3               TRUE 1282.3330 167.49998            #STAR.splice_junctions
4               TRUE 1394.4164 245.74996              #STAR.reads_per_gene
5               TRUE 1505.0830 322.99995                   #STAR.log_files
6               TRUE 1278.7498 446.74996          #STAR.chimeric_junctions</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># flow inputs</span>
f1<span class="op">$</span><span class="kw">input_type</span>()</code></pre></div>
<pre><code>                   sjdbGTFfile                          fastq 
                     "File..."                      "File..." 
              genomeFastaFiles sjdbGTFtagExonParentTranscript 
                        "File"                       "string" 
      sjdbGTFtagExonParentGene          winAnchorMultimapNmax 
                      "string"                          "int" 
            winAnchorDistNbins 
                         "int" </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># flow outouts</span>
f1<span class="op">$</span><span class="kw">output_type</span>()</code></pre></div>
<pre><code>             unmapped_reads transcriptome_aligned_reads 
                  "File..."                      "File" 
           splice_junctions              reads_per_gene 
                     "File"                      "File" 
                  log_files          chimeric_junctions 
                  "File..."                      "File" 
        intermediate_genome         chimeric_alignments 
                     "File"                      "File" 
                 sorted_bam                      result 
                     "File"                      "File" </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># list tools</span>
f1<span class="op">$</span><span class="kw">list_tool</span>()</code></pre></div>
<pre><code>                       label
1       STAR Genome Generate
2 SBG FASTQ Quality Detector
3             Picard SortSam
4                       STAR
                                                  sbgid
1       sevenbridges/public-apps/star-genome-generate/1
2 sevenbridges/public-apps/sbg-fastq-quality-detector/3
3       sevenbridges/public-apps/picard-sortsam-1-140/2
4                       sevenbridges/public-apps/star/4
                           id
1       #STAR_Genome_Generate
2 #SBG_FASTQ_Quality_Detector
3             #Picard_SortSam
4                       #STAR</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># f1$get_tool("STAR")</span></code></pre></div>
<p>There are more utilities please check example at <code>help(Flow)</code></p>
</div>
<div id="create-your-own-flow-in-r" class="section level2">
<h2 class="hasAnchor">
<a href="#create-your-own-flow-in-r" class="anchor"></a>Create your own flow in R</h2>
<div id="introduction-1" class="section level3">
<h3 class="hasAnchor">
<a href="#introduction-1" class="anchor"></a>Introduction</h3>
<p>To create a workflow, we provide simple interface to pipe your tool into a single workflow, it works under situation like</p>
<ul>
<li>Simple linear tool connection and chaining</li>
<li>Connect flow with one or more tools</li>
</ul>
<p><strong>Note</strong> for complicated workflow construction, I highly recommend just use our graphical interface to do it, there is no better way.</p>
</div>
<div id="connect-simple-linear-tools" class="section level3">
<h3 class="hasAnchor">
<a href="#connect-simple-linear-tools" class="anchor"></a>Connect simple linear tools</h3>
<p>Let’s create tools from scratch to perform a simple task</p>
<ol style="list-style-type: decimal">
<li>Tool 1 output 1000 random number</li>
<li>Tool 2 take log on it</li>
<li>Tool 3 do a mean calculation of everything</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"sevenbridges"</span>)
<span class="co"># A tool that generates 100 random numbers</span>
t1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Tool-class.html">Tool</a></span>(<span class="dt">id =</span> <span class="st">"runif new test 3"</span>, <span class="dt">label =</span> <span class="st">"random number"</span>,
           <span class="dt">hints =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(<span class="kw"><a href="../reference/requirements.html">docker</a></span>(<span class="dt">pull =</span> <span class="st">"rocker/r-base"</span>)),
           <span class="dt">baseCommand =</span> <span class="st">"Rscript -e 'x = runif(100); write.csv(x, file = 'random.txt', row.names = FALSE)'"</span>,
           <span class="dt">outputs =</span> <span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"random"</span>,
                            <span class="dt">type =</span> <span class="st">"file"</span>,
                            <span class="dt">glob =</span> <span class="st">"random.txt"</span>))

<span class="co"># A tool that takes log</span>
fd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/requirements.html">fileDef</a></span>(<span class="dt">name =</span> <span class="st">"log.R"</span>,
              <span class="dt">content =</span> <span class="st">"args = commandArgs(TRUE)</span>
<span class="st">                         x = read.table(args[1], header = TRUE)[,'x']</span>
<span class="st">                         x = log(x)</span>
<span class="st">                         write.csv(x, file = 'random_log.txt', row.names = FALSE)</span>
<span class="st">                         "</span>)

t2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Tool-class.html">Tool</a></span>(<span class="dt">id =</span> <span class="st">"log new test 3"</span>, <span class="dt">label =</span> <span class="st">"get log"</span>,
           <span class="dt">hints =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(<span class="kw"><a href="../reference/requirements.html">docker</a></span>(<span class="dt">pull =</span> <span class="st">"rocker/r-base"</span>)),
           <span class="dt">requirements =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(fd),
           <span class="dt">baseCommand =</span> <span class="st">"Rscript log.R"</span>,
           <span class="dt">inputs =</span> <span class="kw"><a href="../reference/shorthand-cwl.html">input</a></span>(<span class="dt">id =</span> <span class="st">"number"</span>,
                           <span class="dt">type =</span> <span class="st">"file"</span>),
           <span class="dt">outputs =</span> <span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"log"</span>,
                            <span class="dt">type =</span> <span class="st">"file"</span>,
                            <span class="dt">glob =</span> <span class="st">"*.txt"</span>))

<span class="co"># A tool that do a mean</span>
fd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/requirements.html">fileDef</a></span>(<span class="dt">name =</span> <span class="st">"mean.R"</span>,
              <span class="dt">content =</span> <span class="st">"args = commandArgs(TRUE)</span>
<span class="st">                         x = read.table(args[1], header = TRUE)[,'x']</span>
<span class="st">                         x = mean(x)</span>
<span class="st">                         write.csv(x, file = 'random_mean.txt', row.names = FALSE)"</span>)

t3 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Tool-class.html">Tool</a></span>(<span class="dt">id =</span> <span class="st">"mean new test 3"</span>, <span class="dt">label =</span> <span class="st">"get mean"</span>,
           <span class="dt">hints =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(<span class="kw"><a href="../reference/requirements.html">docker</a></span>(<span class="dt">pull =</span> <span class="st">"rocker/r-base"</span>)),
           <span class="dt">requirements =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(fd),
           <span class="dt">baseCommand =</span> <span class="st">"Rscript mean.R"</span>,
           <span class="dt">inputs =</span> <span class="kw"><a href="../reference/shorthand-cwl.html">input</a></span>(<span class="dt">id =</span> <span class="st">"number"</span>,
                           <span class="dt">type =</span> <span class="st">"file"</span>),
           <span class="dt">outputs =</span> <span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"mean"</span>,
                            <span class="dt">type =</span> <span class="st">"file"</span>,
                            <span class="dt">glob =</span> <span class="st">"*.txt"</span>))

f =<span class="st"> </span>t1 <span class="op">%&gt;&gt;%</span><span class="st"> </span>t2</code></pre></div>
<pre><code>flow_output: #get_log.log</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">f =<span class="st"> </span><span class="kw"><a href="../reference/link.html">link</a></span>(t1, t2, <span class="st">"#random"</span>, <span class="st">"#number"</span>)</code></pre></div>
<pre><code>flow_output: #get_log.log</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># # you cannot directly copy-paste it</span>
<span class="co"># # please push it using API, we will register each tool for you</span>

<span class="co"># library("clipr")</span>
<span class="co"># write_clip(f$toJSON(pretty = TRUE))</span>

t2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Tool-class.html">Tool</a></span>(<span class="dt">id =</span> <span class="st">"log new test 3"</span>, <span class="dt">label =</span> <span class="st">"get log"</span>,
           <span class="dt">hints =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(<span class="kw"><a href="../reference/requirements.html">docker</a></span>(<span class="dt">pull =</span> <span class="st">"rocker/r-base"</span>)),
           <span class="co"># requirements = requirements(fd),</span>
           <span class="dt">baseCommand =</span> <span class="st">"Rscript log.R"</span>,
           <span class="dt">inputs =</span> <span class="kw"><a href="../reference/shorthand-cwl.html">input</a></span>(<span class="dt">id =</span> <span class="st">"number"</span>,
                           <span class="dt">type =</span> <span class="st">"file"</span>,
                          <span class="dt">secondaryFiles =</span> sevenbridges<span class="op">:::</span><span class="kw">set_box</span>(<span class="st">".bai"</span>)),
           <span class="dt">outputs =</span> <span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"log"</span>,
                            <span class="dt">type =</span> <span class="st">"file"</span>,
                            <span class="dt">glob =</span> <span class="st">"*.txt"</span>))

<span class="co"># library("clipr")</span>
<span class="co"># write_clip(t2$toJSON(pretty = TRUE))</span></code></pre></div>
<p><strong>Note</strong>: this workflow contains tools that do not exist on the platform, so if you directly copy and paste the JSON into the GUI, it won’t work properly, however, a simple way is to push your app to platform via API. This will add new tools one by one to your project before add your workflow app on the platform. Alternative if you connect two tools you know they exist on the platform, you don’t need to do so.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># auto-check tool info and push new tools</span>
p<span class="op">$</span><span class="kw">app_add</span>(<span class="st">"new_flow_log"</span>, f)</code></pre></div>
</div>
<div id="connecting-tools-by-input-and-output-id" class="section level3">
<h3 class="hasAnchor">
<a href="#connecting-tools-by-input-and-output-id" class="anchor"></a>Connecting tools by input and output id</h3>
<p>Now let’s connect two tools</p>
<ol style="list-style-type: decimal">
<li>Unpakcing a compressed fastq files</li>
<li>STAR aligner</li>
</ol>
<p>Checking potential mapping is easy with function <code>link_what</code>, it will print matched input and outputs. Then the generic function <code>link</code> will allow you to connect two <code>Tool</code> objects</p>
<p>If you don’t specify which input/ouput to expose at flow level for new <code>Flow</code> object, it will expose all availabl ones and print the message, otherwise, please provide parameters for <code>flow_input</code> and <code>flow_output</code> with full id.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">t1 =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/app"</span>, <span class="st">"tool_unpack_fastq.json"</span>,
                 <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
t2 =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/app"</span>, <span class="st">"tool_star.json"</span>,
                 <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
t1 =<span class="st"> </span><span class="kw"><a href="../reference/convert_app.html">convert_app</a></span>(t1)
t2 =<span class="st"> </span><span class="kw"><a href="../reference/convert_app.html">convert_app</a></span>(t2)
<span class="co"># check possible link</span>
<span class="kw"><a href="../reference/link_what.html">link_what</a></span>(t1, t2)</code></pre></div>
<pre><code>$File...
$File...$from
                   id              label    type fileTypes
1 #output_fastq_files Output FASTQ files File...     FASTQ
           full.name
1 #SBG_Unpack_FASTQs

$File...$to
             id                label    type required prefix
1        #reads        Read sequence File...     TRUE   &lt;NA&gt;
95 #sjdbGTFfile Splice junction file File...    FALSE   &lt;NA&gt;
                                                  fileTypes full.name
1  FASTA, FASTQ, FA, FQ, FASTQ.GZ, FQ.GZ, FASTQ.BZ2, FQ.BZ2     #STAR
95                                            GTF, GFF, TXT     #STAR</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># link</span>
f1 =<span class="st"> </span><span class="kw"><a href="../reference/link.html">link</a></span>(t1, t2, <span class="st">"output_fastq_files"</span>, <span class="st">"reads"</span>)</code></pre></div>
<pre><code>flow_input: #SBG_Unpack_FASTQs.input_archive_file / #STAR.sjdbGTFfile / #STAR.genome</code></pre>
<pre><code>flow_output: #STAR.aligned_reads / #STAR.transcriptome_aligned_reads / #STAR.reads_per_gene / #STAR.log_files / #STAR.splice_junctions / #STAR.chimeric_junctions / #STAR.unmapped_reads / #STAR.intermediate_genome / #STAR.chimeric_alignments</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># link</span>
t1<span class="op">$</span><span class="kw">output_id</span>(<span class="ot">TRUE</span>)</code></pre></div>
<pre><code>                                File... 
"#SBG_Unpack_FASTQs.output_fastq_files" </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">t2<span class="op">$</span><span class="kw">input_id</span>(<span class="ot">TRUE</span>)</code></pre></div>
<pre><code>                                 File... 
                           "#STAR.reads" 
                                    enum 
              "#STAR.readMatesLengthsIn" 
                                     int 
                   "#STAR.readMapNumber" 
                                     int 
               "#STAR.limitOutSJoneRead" 
                                     int 
             "#STAR.limitOutSJcollapsed" 
                                    enum 
                "#STAR.outReadsUnmapped" 
                                     int 
              "#STAR.outQSconversionAdd" 
                                    enum 
                      "#STAR.outSAMtype" 
                                    enum 
                  "#STAR.outSortingType" 
                                    enum 
                      "#STAR.outSAMmode" 
                                    enum 
               "#STAR.outSAMstrandField" 
                                    enum 
                "#STAR.outSAMattributes" 
                                    enum 
                  "#STAR.outSAMunmapped" 
                                    enum 
                     "#STAR.outSAMorder" 
                                    enum 
               "#STAR.outSAMprimaryFlag" 
                                    enum 
                    "#STAR.outSAMreadID" 
                                     int 
                "#STAR.outSAMmapqUnique" 
                                     int 
                    "#STAR.outSAMflagOR" 
                                     int 
                   "#STAR.outSAMflagAND" 
                                  string 
                  "#STAR.outSAMheaderHD" 
                                  string 
                  "#STAR.outSAMheaderPG" 
                                  string 
                   "#STAR.rg_seq_center" 
                                  string 
                   "#STAR.rg_library_id" 
                                  string 
                          "#STAR.rg_mfl" 
                                    enum 
                     "#STAR.rg_platform" 
                                  string 
             "#STAR.rg_platform_unit_id" 
                                  string 
                    "#STAR.rg_sample_id" 
                                    enum 
                   "#STAR.outFilterType" 
                                     int 
     "#STAR.outFilterMultimapScoreRange" 
                                     int 
           "#STAR.outFilterMultimapNmax" 
                                     int 
           "#STAR.outFilterMismatchNmax" 
                                   float 
      "#STAR.outFilterMismatchNoverLmax" 
                                   float 
  "#STAR.outFilterMismatchNoverReadLmax" 
                                     int 
               "#STAR.outFilterScoreMin" 
                                   float 
      "#STAR.outFilterScoreMinOverLread" 
                                     int 
              "#STAR.outFilterMatchNmin" 
                                   float 
     "#STAR.outFilterMatchNminOverLread" 
                                    enum 
           "#STAR.outFilterIntronMotifs" 
                                    enum 
                "#STAR.outSJfilterReads" 
                                  int... 
          "#STAR.outSJfilterOverhangMin" 
                                  int... 
       "#STAR.outSJfilterCountUniqueMin" 
                                  int... 
        "#STAR.outSJfilterCountTotalMin" 
                                  int... 
     "#STAR.outSJfilterDistToOtherSJmin" 
                                  int... 
     "#STAR.outSJfilterIntronMaxVsReadN" 
                                     int 
                        "#STAR.scoreGap" 
                                     int 
                  "#STAR.scoreGapNoncan" 
                                     int 
                    "#STAR.scoreGapGCAG" 
                                     int 
                    "#STAR.scoreGapATAC" 
                                   float 
     "#STAR.scoreGenomicLengthLog2scale" 
                                     int 
                    "#STAR.scoreDelOpen" 
                                     int 
                    "#STAR.scoreDelBase" 
                                     int 
                    "#STAR.scoreInsOpen" 
                                     int 
                    "#STAR.scoreInsBase" 
                                     int 
              "#STAR.scoreStitchSJshift" 
                                     int 
             "#STAR.seedSearchStartLmax" 
                                   float 
    "#STAR.seedSearchStartLmaxOverLread" 
                                     int 
                  "#STAR.seedSearchLmax" 
                                     int 
                "#STAR.seedMultimapNmax" 
                                     int 
                 "#STAR.seedPerReadNmax" 
                                     int 
               "#STAR.seedPerWindowNmax" 
                                     int 
           "#STAR.seedNoneLociPerWindow" 
                                     int 
                  "#STAR.alignIntronMin" 
                                     int 
                  "#STAR.alignIntronMax" 
                                     int 
                "#STAR.alignMatesGapMax" 
                                     int 
              "#STAR.alignSJoverhangMin" 
                                     int 
            "#STAR.alignSJDBoverhangMin" 
                                     int 
         "#STAR.alignSplicedMateMapLmin" 
                                   float 
"#STAR.alignSplicedMateMapLminOverLmate" 
                                   float 
         "#STAR.alignWindowsPerReadNmax" 
                                     int 
   "#STAR.alignTranscriptsPerWindowNmax" 
                                     int 
     "#STAR.alignTranscriptsPerReadNmax" 
                                    enum 
                   "#STAR.alignEndsType" 
                                    enum 
    "#STAR.alignSoftClipAtReferenceEnds" 
                                     int 
           "#STAR.winAnchorMultimapNmax" 
                                     int 
                     "#STAR.winBinNbits" 
                                     int 
              "#STAR.winAnchorDistNbins" 
                                     int 
                   "#STAR.winFlankNbins" 
                                     int 
                  "#STAR.chimSegmentMin" 
                                     int 
                    "#STAR.chimScoreMin" 
                                     int 
                "#STAR.chimScoreDropMax" 
                                     int 
             "#STAR.chimScoreSeparation" 
                                     int 
        "#STAR.chimScoreJunctionNonGTAG" 
                                     int 
         "#STAR.chimJunctionOverhangMin" 
                                    enum 
                       "#STAR.quantMode" 
                                     int 
                  "#STAR.twopass1readsN" 
                                    enum 
                     "#STAR.twopassMode" 
                                  string 
                   "#STAR.genomeDirName" 
                                    enum 
                  "#STAR.sjdbInsertSave" 
                                  string 
                "#STAR.sjdbGTFchrPrefix" 
                                  string 
              "#STAR.sjdbGTFfeatureExon" 
                                  string 
  "#STAR.sjdbGTFtagExonParentTranscript" 
                                  string 
        "#STAR.sjdbGTFtagExonParentGene" 
                                     int 
                    "#STAR.sjdbOverhang" 
                                     int 
                       "#STAR.sjdbScore" 
                                 File... 
                     "#STAR.sjdbGTFfile" 
                                  int... 
                    "#STAR.clip3pNbases" 
                                  int... 
                    "#STAR.clip5pNbases" 
                               string... 
                "#STAR.clip3pAdapterSeq" 
                                float... 
                "#STAR.clip3pAdapterMMp" 
                                  int... 
        "#STAR.clip3pAfterAdapterNbases" 
                                    enum 
                     "#STAR.chimOutType" 
                                    File 
                          "#STAR.genome" 
                                     int 
              "#STAR.limitSjdbInsertNsj" 
                                    enum 
           "#STAR.quantTranscriptomeBan" 
                                     int 
                 "#STAR.limitBAMsortRAM" </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">f2 =<span class="st"> </span><span class="kw"><a href="../reference/link.html">link</a></span>(t1, t2, <span class="st">"output_fastq_files"</span>, <span class="st">"reads"</span>,
          <span class="dt">flow_input =</span> <span class="st">"#SBG_Unpack_FASTQs.input_archive_file"</span>,
          <span class="dt">flow_output =</span> <span class="st">"#STAR.log_files"</span>)</code></pre></div>
<pre><code>flow_input: #SBG_Unpack_FASTQs.input_archive_file / #STAR.genome</code></pre>
<pre><code>flow_output: #STAR.log_files</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># library("clipr")</span>
<span class="co"># write_clip(f2$toJSON())</span></code></pre></div>
</div>
<div id="connecting-tool-with-workflow-by-input-and-output-id" class="section level3">
<h3 class="hasAnchor">
<a href="#connecting-tool-with-workflow-by-input-and-output-id" class="anchor"></a>Connecting tool with workflow by input and output id</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tool.in =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/app"</span>, <span class="st">"tool_unpack_fastq.json"</span>, <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
flow.in =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/app"</span>, <span class="st">"flow_star.json"</span>, <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)

t1 =<span class="st"> </span><span class="kw"><a href="../reference/convert_app.html">convert_app</a></span>(tool.in)
f2 =<span class="st"> </span><span class="kw"><a href="../reference/convert_app.html">convert_app</a></span>(flow.in)
<span class="co"># consulting link_what first</span>
f2<span class="op">$</span><span class="kw">link_map</span>()</code></pre></div>
<pre><code>                                                     id
1  #STAR_Genome_Generate.sjdbGTFtagExonParentTranscript
2        #STAR_Genome_Generate.sjdbGTFtagExonParentGene
3                     #STAR_Genome_Generate.sjdbGTFfile
4                #STAR_Genome_Generate.genomeFastaFiles
5                     #SBG_FASTQ_Quality_Detector.fastq
6                             #Picard_SortSam.input_bam
7                           #STAR.winAnchorMultimapNmax
8                              #STAR.winAnchorDistNbins
9                                     #STAR.sjdbGTFfile
10                                          #STAR.reads
11                                         #STAR.genome
12                                      #unmapped_reads
13                         #transcriptome_aligned_reads
14                                    #splice_junctions
15                                      #reads_per_gene
16                                           #log_files
17                                  #chimeric_junctions
18                                 #intermediate_genome
19                                 #chimeric_alignments
20                                          #sorted_bam
21                                              #result
                               source   type
1     #sjdbGTFtagExonParentTranscript  input
2           #sjdbGTFtagExonParentGene  input
3                        #sjdbGTFfile  input
4                   #genomeFastaFiles  input
5                              #fastq  input
6                 #STAR.aligned_reads  input
7              #winAnchorMultimapNmax  input
8                 #winAnchorDistNbins  input
9                        #sjdbGTFfile  input
10 #SBG_FASTQ_Quality_Detector.result  input
11       #STAR_Genome_Generate.genome  input
12               #STAR.unmapped_reads output
13  #STAR.transcriptome_aligned_reads output
14             #STAR.splice_junctions output
15               #STAR.reads_per_gene output
16                    #STAR.log_files output
17           #STAR.chimeric_junctions output
18          #STAR.intermediate_genome output
19          #STAR.chimeric_alignments output
20         #Picard_SortSam.sorted_bam output
21 #SBG_FASTQ_Quality_Detector.result output</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># then link</span>

f3 =<span class="st"> </span><span class="kw"><a href="../reference/link.html">link</a></span>(t1, f2, <span class="kw">c</span>(<span class="st">"output_fastq_files"</span>), <span class="kw">c</span>(<span class="st">"#SBG_FASTQ_Quality_Detector.fastq"</span>))

<span class="kw"><a href="../reference/link_what.html">link_what</a></span>(f2, t1)</code></pre></div>
<pre><code>$File
$File$from
                             id                       label type required
2  #transcriptome_aligned_reads transcriptome_aligned_reads File    FALSE
3             #splice_junctions            splice_junctions File    FALSE
4               #reads_per_gene              reads_per_gene File    FALSE
6           #chimeric_junctions          chimeric_junctions File    FALSE
7          #intermediate_genome         intermediate_genome File    FALSE
8          #chimeric_alignments         chimeric_alignments File    FALSE
9                   #sorted_bam                  sorted_bam File    FALSE
10                      #result                      result File    FALSE
   fileTypes                            link_to
2       null  #STAR.transcriptome_aligned_reads
3       null             #STAR.splice_junctions
4       null               #STAR.reads_per_gene
6       null           #STAR.chimeric_junctions
7       null          #STAR.intermediate_genome
8       null          #STAR.chimeric_alignments
9       null         #Picard_SortSam.sorted_bam
10      null #SBG_FASTQ_Quality_Detector.result

$File$to
                   id              label type required
1 #input_archive_file Input archive file File     TRUE
                prefix                                      fileTypes
1 --input_archive_file TAR, TAR.GZ, TGZ, TAR.BZ2, TBZ2,  GZ, BZ2, ZIP</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">f4 =<span class="st"> </span><span class="kw"><a href="../reference/link.html">link</a></span>(f2, t1, <span class="kw">c</span>(<span class="st">"#Picard_SortSam.sorted_bam"</span>, <span class="st">"#SBG_FASTQ_Quality_Detector.result"</span>), <span class="kw">c</span>(<span class="st">"#input_archive_file"</span>, <span class="st">"#input_archive_file"</span>))</code></pre></div>
<pre><code>flow_input: #SBG_Unpack_FASTQs.input_archive_file</code></pre>
<pre><code>flow_output: #SBG_Unpack_FASTQs.output_fastq_files</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># # </span><span class="al">TODO</span>
<span class="co"># # all outputs</span>
<span class="co"># # flow + flow</span>
<span class="co"># # print message when name wrong</span>
<span class="co"># library("clipr")</span>
<span class="co"># write_clip(f4$toJSON())</span></code></pre></div>
</div>
<div id="using-pipe-to-construct-complicated-workflow" class="section level3">
<h3 class="hasAnchor">
<a href="#using-pipe-to-construct-complicated-workflow" class="anchor"></a>Using pipe to construct complicated workflow</h3>
</div>
</div>
</div>
<div id="execution" class="section level1">
<h1 class="hasAnchor">
<a href="#execution" class="anchor"></a>Execution</h1>
<div id="execute-the-tool-and-flow-in-the-cloud" class="section level2">
<h2 class="hasAnchor">
<a href="#execute-the-tool-and-flow-in-the-cloud" class="anchor"></a>Execute the tool and flow in the cloud</h2>
<p>With API function, you can directly load your Tool into the account. Run a task, for “how-to”, please check the complete guide for API client.</p>
<p>Here is a quick demo:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a =<span class="st"> </span><span class="kw"><a href="../reference/Auth-class.html">Auth</a></span>(<span class="dt">platform =</span> <span class="st">"platform_name"</span>, <span class="dt">token =</span> <span class="st">"your_token"</span>)
p =<span class="st"> </span>a<span class="op">$</span><span class="kw">project</span>(<span class="st">"demo"</span>)
app.runif =<span class="st"> </span>p<span class="op">$</span><span class="kw">app_add</span>(<span class="st">"runif555"</span>, rbx)
aid =<span class="st"> </span>app.runif<span class="op">$</span>id
tsk =<span class="st"> </span>p<span class="op">$</span><span class="kw">task_add</span>(<span class="dt">name        =</span> <span class="st">"Draft runif simple"</span>,
                 <span class="dt">description =</span> <span class="st">"Description for runif"</span>,
                 <span class="dt">app         =</span> aid,
                 <span class="dt">inputs      =</span> <span class="kw">list</span>(<span class="dt">min =</span> <span class="dv">1</span>, <span class="dt">max =</span> <span class="dv">10</span>))
tsk<span class="op">$</span><span class="kw">run</span>()</code></pre></div>
</div>
<div id="execute-the-tool-in-rabix-test-locally" class="section level2">
<h2 class="hasAnchor">
<a href="#execute-the-tool-in-rabix-test-locally" class="anchor"></a>Execute the tool in Rabix – test locally</h2>
<p><strong>1. From CLI</strong></p>
<p>While developing tools it is useful to test them locally first. For that we can use rabix – reproducible analyses for bioinformatics, <a href="https://github.com/rabix" class="uri">https://github.com/rabix</a>. To test your tool with latest implementation of rabix in Java (called <strong>bunny</strong>) you could use the Docker image <strong>tengfei/testenv</strong>:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">docker</span> pull tengfei/testenv</code></pre></div>
<p>Dump your rabix tool as JSON into dir which also contains input data. <code>write(rbx$toJSON, file="&lt;data_dir&gt;/&lt;tool&gt;.json")</code>. Make <strong>inputs.json</strong> file to declare input parameters in the same directory (you can use relative paths from inputs.json to data). Create container:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">docker</span> run --privileged --name bunny -v <span class="op">&lt;</span>/path/to/data_dir<span class="op">&gt;</span>:/bunny_data -dit tengfei/testenv</code></pre></div>
<p>Execute tool</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">docker</span> exec bunny bash -c <span class="st">'cd /opt/bunny &amp;&amp; ./rabix.sh -e /bunny_data /bunny_data/&lt;tool&gt;.json /bunny_data/inputs.json'</span></code></pre></div>
<p>You’ll see running logs from within container, and also output dir inside <data_dir> in home system.</data_dir></p>
<p>NOTE 1: tengfei/testenv has R, python, Java… so many tools can work without Docker requirement set. If you however set Docker requirement you need to pull image inside container first to run docker container inside running bunny docker.</p>
<p>NOTE 2: inputs.json can also be inputs.yaml if you find it easier to declare inputs in YAML.</p>
<p><strong>2. From R</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"sevenbridges"</span>)

in.df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">id =</span> <span class="kw">c</span>(<span class="st">"number"</span>, <span class="st">"min"</span>, <span class="st">"max"</span>, <span class="st">"seed"</span>),
                    <span class="dt">description =</span> <span class="kw">c</span>(<span class="st">"number of observation"</span>,
                                    <span class="st">"lower limits of the distribution"</span>,
                                    <span class="st">"upper limits of the distribution"</span>,
                                    <span class="st">"seed with set.seed"</span>),
                    <span class="dt">type =</span> <span class="kw">c</span>(<span class="st">"integer"</span>, <span class="st">"float"</span>, <span class="st">"float"</span>, <span class="st">"float"</span>),
                    <span class="dt">label =</span> <span class="kw">c</span>(<span class="st">"number"</span> ,<span class="st">"min"</span>, <span class="st">"max"</span>, <span class="st">"seed"</span>),
                    <span class="dt">prefix =</span> <span class="kw">c</span>(<span class="st">"--n"</span>, <span class="st">"--min"</span>, <span class="st">"--max"</span>, <span class="st">"--seed"</span>),
                    <span class="dt">default =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">10</span>, <span class="dv">123</span>),
                    <span class="dt">required =</span> <span class="kw">c</span>(<span class="ot">TRUE</span>, <span class="ot">FALSE</span>, <span class="ot">FALSE</span>, <span class="ot">FALSE</span>))
out.df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">id =</span> <span class="kw">c</span>(<span class="st">"random"</span>, <span class="st">"report"</span>),
                     <span class="dt">type =</span> <span class="kw">c</span>(<span class="st">"file"</span>, <span class="st">"file"</span>),
                     <span class="dt">glob =</span> <span class="kw">c</span>(<span class="st">"*.txt"</span>, <span class="st">"*.html"</span>))
rbx &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Tool-class.html">Tool</a></span>(<span class="dt">id =</span> <span class="st">"runif"</span>,
            <span class="dt">label =</span> <span class="st">"Random number generator"</span>,
            <span class="dt">hints =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(<span class="kw"><a href="../reference/requirements.html">docker</a></span>(<span class="dt">pull =</span> <span class="st">"tengfei/runif"</span>),
                                 <span class="kw"><a href="../reference/requirements.html">cpu</a></span>(<span class="dv">1</span>), <span class="kw"><a href="../reference/requirements.html">mem</a></span>(<span class="dv">2000</span>)),
            <span class="dt">baseCommand =</span> <span class="st">"runif.R"</span>,
            <span class="dt">inputs =</span> in.df,  <span class="co"># or ins.df</span>
            <span class="dt">outputs =</span> out.df)
params &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">number =</span> <span class="dv">3</span>, <span class="dt">max =</span> <span class="dv">5</span>)

<span class="kw"><a href="../reference/set_test_env.html">set_test_env</a></span>(<span class="st">"tengfei/testenv"</span>, <span class="st">"mount_dir"</span>)
<span class="kw">test_tool</span>(rbx, params)</code></pre></div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#prerequisite">Prerequisite</a></li>
      <li><a href="#app-workflow-and-tool">App, Workflow, and Tool</a></li>
      <li>
<a href="#describe-tools-in-r">Describe Tools in R</a><ul class="nav nav-pills nav-stacked">
<li><a href="#import-from-json-file">Import from JSON file</a></li>
      <li><a href="#utilitites-for-tool-object">Utilitites for Tool object</a></li>
      <li><a href="#create-your-own-tool-in-r">Create your own tool in R</a></li>
      </ul>
</li>
      <li>
<a href="#describe-wokrflow-in-r">Describe Wokrflow in R</a><ul class="nav nav-pills nav-stacked">
<li><a href="#import-from-a-json-file">Import from a JSON file</a></li>
      <li><a href="#utilities-for-flow-objects">Utilities for <code>Flow</code> objects</a></li>
      <li><a href="#create-your-own-flow-in-r">Create your own flow in R</a></li>
      </ul>
</li>
      <li>
<a href="#execution">Execution</a><ul class="nav nav-pills nav-stacked">
<li><a href="#execute-the-tool-and-flow-in-the-cloud">Execute the tool and flow in the cloud</a></li>
      <li><a href="#execute-the-tool-in-rabix-test-locally">Execute the tool in Rabix – test locally</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Nan Xiao, Dusan Randjelovic, Tengfei Yin, Seven Bridges Genomics.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
