<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Master Tutorial: Use R for Cancer Genomics Cloud • sevenbridges</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Google analytics --><script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-87185344-1', 'auto');
  ga('send', 'pageview');

</script>
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">sevenbridges-r</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/api.html">Complete Guide for Seven Bridges API R Client</a>
    </li>
    <li>
      <a href="../articles/apps.html">Describe and Execute CWL Tools/Workflows in R</a>
    </li>
    <li>
      <a href="../articles/bioc-workflow.html">Master Tutorial: Use R for Cancer Genomics Cloud</a>
    </li>
    <li>
      <a href="../articles/cgc-sparql.html">Find Data on CGC via Data Exploerer, SPARQL, and Datasets API</a>
    </li>
    <li>
      <a href="../articles/docker.html">Creating Your Docker Container and Command Line Interface (with docopt)</a>
    </li>
    <li>
      <a href="../articles/rstudio.html">IDE Container: RStudio Server, Shiny Server, and More</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/sbg/sevenbridges-r">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Master Tutorial: Use R for Cancer Genomics Cloud</h1>
                        <h4 class="author">Tengfei Yin &lt;<a href="mailto:tengfei.yin@sevenbridges.com">tengfei.yin@sevenbridges.com</a>&gt;</h4>
            
            <h4 class="date">2018-02-14</h4>
          </div>

    
    
<div class="contents">
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This tutorials originates from 2016 Cancer Genomics Cloud Hackathon R workshop I prepared, and it’s recommended for beginner to read and run through all examples here yourself in your R IDE like Rstudio. Then try to make your own app.</p>
<p>In this tutorial, you will learn:</p>
<ol style="list-style-type: decimal">
<li>API client in R with <strong>sevenbridges</strong> R package to fully automate analysis</li>
<li>Describe command line interface with R package <strong>docopt</strong>
</li>
<li>Make your own Docker app</li>
<li>Describe a standard RNA-seq Bioconductor workflow in CWL with pre-defined report template</li>
<li>Execute it in the cloud</li>
<li>Reporting tool to generate as many R markdown report or Shiny apps report as you want</li>
<li>Deploy directly on shiny server like shinyapps.io from a “report” tool</li>
</ol>
</div>
<div id="prerequisites" class="section level1">
<h1 class="hasAnchor">
<a href="#prerequisites" class="anchor"></a>Prerequisites</h1>
<p>This tutorial doesn’t require you to be an advanced R user, everything you need is R or even better, a cool IDE like Rstudio (or Emacs+ESS), then just open this R Markdown document in Rstudio. It’s easy to learn!</p>
<p>Suggest learning for all users: Docker.</p>
<p>Now we are ready to go!</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>First download the <a href="https://raw.githubusercontent.com/sbg/sevenbridges-r/master/vignettes/bioc-workflow.Rmd">R Markdown source</a> of this page, so you can load it to your RStudio or your favorite IDE to run through all examples and tweak setup.</p>
<p>This package <code>sevenbridges</code> is available on Bioconductor (<a href="https://bioconductor.org/packages/release/bioc/html/sevenbridges.html">release branch</a>, <a href="https://bioconductor.org/packages/devel/bioc/html/sevenbridges.html">development branch</a>). The latest development version is on <a href="https://github.com/sbg/sevenbridges-r">GitHub</a>.</p>
<p>To install the latest development version from GitHub, run the following commands in R:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">if</span> (<span class="op">!</span><span class="kw">require</span>(<span class="st">"devtools"</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)) <span class="kw">install.packages</span>(<span class="st">"devtools"</span>)
<span class="kw">source</span>(<span class="st">"https://bioconductor.org/biocLite.R"</span>)
devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(
    <span class="st">"sbg/sevenbridges-r"</span>, <span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>,
    <span class="dt">repos =</span> BiocInstaller<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/BiocInstaller/topics/biocinstallRepos">biocinstallRepos</a></span>(), <span class="dt">dependencies =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>After the installation you can always browse the vignettes</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">browseVignettes</span>(<span class="dt">package =</span> <span class="st">"sevenbridges"</span>)</code></pre></div>
</div>
<div id="register-on-cancer-genomics-cloud" class="section level2">
<h2 class="hasAnchor">
<a href="#register-on-cancer-genomics-cloud" class="anchor"></a>Register on Cancer Genomics Cloud</h2>
<p><img src="https://i.imgur.com/p52wIsi.png" width="100%" class="img-responsive" alt="cgc-home"></p>
<p>You can find login/registration on the Cancer Genomics Cloud homepage <a href="https://www.cancergenomicscloud.org" class="uri">https://www.cancergenomicscloud.org</a>. Follow the <a href="http://docs.cancergenomicscloud.org/docs/sign-up-for-the-cgc">signup tutorial</a> if you need to access TCGA Controlled Data on the CGC via NIH eRA Commons.</p>
</div>
<div id="authentication" class="section level2">
<h2 class="hasAnchor">
<a href="#authentication" class="anchor"></a>Authentication</h2>
<p>After logged in, you can get your authentication token under your account setting and the “Developer” tab (<a href="http://docs.cancergenomicscloud.org/docs/get-your-authentication-token">tutorial</a>).</p>
</div>
<div id="register-on-shinyapps-io-optional" class="section level2">
<h2 class="hasAnchor">
<a href="#register-on-shinyapps-io-optional" class="anchor"></a>Register on shinyapps.io (Optional)</h2>
<p>In this tutorial, if you want to try to deploy the Shiny web application automatically on a remote server like shinyapps.io, please visit <a href="https://www.shinyapps.io/" class="uri">https://www.shinyapps.io/</a> to register and login.</p>
<p>Get you token and secret ready to deploy:</p>
<p><img src="https://i.imgur.com/ChCBhf6.png" width="100%" class="img-responsive" alt="shinyapps-token"></p>
</div>
<div id="report-issues" class="section level2">
<h2 class="hasAnchor">
<a href="#report-issues" class="anchor"></a>Report issues</h2>
<p>This package is under active development, will bring many new features as well, at any moment, if you have questions or problem about this R package, please file issues on <a href="https://github.com/sbg/sevenbridges-r/issues">GitHub</a>.</p>
<p>If you have question regarding the Cancer Genomics Cloud or other Seven Bridges platforms, we have a different channel for each platform, for example, Cancer Genomics Cloud have lots <a href="http://docs.cancergenomicscloud.org/docs">documentation</a> and a <a href="http://docs.cancergenomicscloud.org/discuss">forum</a>.</p>
<p>Please, feedback is always welcomed!</p>
</div>
</div>
<div id="quickstart" class="section level1">
<h1 class="hasAnchor">
<a href="#quickstart" class="anchor"></a>Quickstart</h1>
<p>The final goal is make a workflow that</p>
<ol style="list-style-type: decimal">
<li>Input gene feature, design matrix, bam files, and generate differential expression report and output full report, a picture and a count table as example.</li>
<li>Add report tool with two Shiny app template and two R Markdown template to collect files from previous flow and generate new report, even deploy on shinyapps.io automatically after a task is finished.</li>
</ol>
<p>The final workflow looks like this, it’s composed of two tools: RNA-seq analysis tool and reporting tool.</p>
<p><img src="https://i.imgur.com/OkgB2tl.png" width="100%" class="img-responsive" alt="quickstart-flow"></p>
<p>The Shiny app report with ggvis module on the shinyapps.io server looks like <a href="https://tengfei.shinyapps.io/scatter_plot/">this</a></p>
<p>A ggvis interactive scatter plot</p>
<p><img src="https://i.imgur.com/X8M2O0T.png" width="100%" class="img-responsive" alt="ggvis"></p>
<p>A differential expression table</p>
<p><img src="https://i.imgur.com/wQ9O3Ys.png" width="100%" class="img-responsive" alt="de-table"></p>
<p>A full HTML report included, it’s also the output from the first tool, in this way, you can orchestrate many tools output into single report for your task.</p>
<p><img src="https://i.imgur.com/9vgBQQC.png" width="100%" class="img-responsive" alt="html-report"></p>
<p><img src="https://i.imgur.com/l5Xgq9R.png" width="100%" class="img-responsive" alt="heatmap"></p>
<p>Now let’s start building tools.</p>
<div id="create-a-project-under-your-account-via-api-r-client" class="section level2">
<h2 class="hasAnchor">
<a href="#create-a-project-under-your-account-via-api-r-client" class="anchor"></a>Create a project under your account via API R client</h2>
<p>I know, we can always do it via graphic user interface, but let’s have fun with the <code>sevenbridges</code> packages you just installed.</p>
<p>For complete API tutorial and reference manual, please read another tutorial.</p>
<pre><code>vignette("api", package = "sevenbridges")</code></pre>
<p>Now let’s do some simple steps, first thing to do is to create an <em>Auth</em> object, almost everything started from this object. Our API client follow a style like this <code>Auth$properties$action</code>. On the platform, <code>Auth</code> is your account, and it contains projects, billing groups, users, project contains tasks, apps, files etc, so it’s easy to imagine your API call.</p>
<p>To create an <code>Auth</code> object, simply pass the token and platform name (or alternatively, API base URL) to the <code>Auth</code> function. The default platform is set to CGC. Good news you can use the <code>sevenbridges</code> package to access any Seven Bridges platform with API (v2).</p>
<p>This is the main way to create an Auth object, just replace <code>your_token</code> with your own authentication token from the CGC:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Auth-class.html">Auth</a></span>(<span class="dt">token =</span> <span class="st">"your_token"</span>, <span class="dt">platform =</span> <span class="st">"cgc"</span>)</code></pre></div>
<p>Alternatively, you can save your authentication cross different platforms in a user configuration file named <code>credentials</code> under the directory <code>$HOME/.sevenbridges/</code>. This allows you to manage multiple user profiles registered on multiple Seven Bridges environments. An example user configuration file looks like this:</p>
<pre><code>[aws-us-tengfei]
api_endpoint = https://api.sbgenomics.com/v2
auth_token = token_for_this_user

# This is a comment:
# another user on platform aws-us
[aws-us-yintengfei]
api_endpoint = https://api.sbgenomics.com/v2
auth_token = token_for_this_user

[cgc]
api_endpoint = https://cgc-api.sbgenomics.com/v2
auth_token = token_for_this_user

[gcp]
api_endpoint = https://gcp-api.sbgenomics.com/v2
auth_token = token_for_this_user</code></pre>
<p>When you have this user configuration file ready at this default location, all you need to do is setting <code>from = "file"</code> and choose the <code>profile_name</code> to use. For example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Auth-class.html">Auth</a></span>(<span class="dt">from =</span> <span class="st">"file"</span>, <span class="dt">profile_name =</span> <span class="st">"cgc"</span>)
<span class="co"># # remove old project</span>
<span class="co"># a$project(id = "tengfei/hackathon")$delete()</span></code></pre></div>
<p>The third way to save authentication information is by setting system environment variables. To set the two environment variables (<code>SB_API_ENDPOINT</code> and <code>SB_AUTH_TOKEN</code>) in your system, you could use the function <code><a href="../reference/sbg_set_env.html">sbg_set_env()</a></code>. For example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/sbg_set_env.html">sbg_set_env</a></span>(<span class="st">"https://cgc-api.sbgenomics.com/v2"</span>, <span class="st">"your_token"</span>)</code></pre></div>
<p>To create an <code>Auth</code> object using credentials in the environment variables:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Auth-class.html">Auth</a></span>(<span class="dt">from =</span> <span class="st">"env"</span>)</code></pre></div>
<p>To create a new project, you need to know your billing group id, cost related to this project will be charged from this billing group, now play with your free credit.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">(b &lt;-<span class="st"> </span>a<span class="op">$</span><span class="kw">billing</span>())
<span class="co"># a single billing group is showing</span></code></pre></div>
<p>Now let’s create a new project called “hackathon”, save it to an object named <code>p</code> for convenient usage for any call related to this project.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">(p &lt;-<span class="st"> </span>a<span class="op">$</span><span class="kw">project_new</span>(<span class="st">"hackathon"</span>,
                    <span class="dt">billing_group_id =</span> b<span class="op">$</span>id,
                    <span class="dt">description =</span> <span class="st">"This project is for CGC hackathon"</span>))</code></pre></div>
<p>Now check it on CGC, you will see a fresh new project is created.</p>
<p><img src="https://i.imgur.com/oI0dXyu.png" width="100%" class="img-responsive" alt="cgc-projects"></p>
<p>To delete it, just call, but I will suggest you keep it for following tutorial : )</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># p$delete()</span></code></pre></div>
</div>
<div id="build-a-rna-seq-tool-from-bam-to-report" class="section level2">
<h2 class="hasAnchor">
<a href="#build-a-rna-seq-tool-from-bam-to-report" class="anchor"></a>Build a RNA-Seq tool: from bam to report</h2>
<div id="step-1-have-a-plan" class="section level3">
<h3 class="hasAnchor">
<a href="#step-1-have-a-plan" class="anchor"></a>Step 1: Have a plan</h3>
<p>To demonstrate I will use a RNA-seq <a href="https://www.bioconductor.org/help/workflows/rnaseqGene/">workflow</a> from Bioconductor, when you open this link, you will notice a full example written in Markdown I want to make it into an app and allow you to input new files and generate new report base on this template.</p>
<p>Before you build any tool, you have to have a plan in mind</p>
<ol style="list-style-type: decimal">
<li>What Input, Output and Parameter you want to define for the tool</li>
<li>Is there a Docker container already available for your tool? Do you want to build one with command line interface?</li>
</ol>
<p>First thing first, let’s define our tool first, here is a diagram generated with Seven Bridges’ graphical user interface.</p>
<p><img src="fig/diagram_rnaseq.png" width="100%" class="img-responsive" alt="rnaseq-diagram"></p>
<ul>
<li>Input
<ul>
<li>Gene feature file</li>
<li>A list of bamfiles</li>
<li>Design matrix</li>
</ul>
</li>
<li>Output
<ul>
<li>Report in PDF, using the R Markdown associated with this workflow and example</li>
<li>Graphics</li>
<li>Differential expression table</li>
</ul>
</li>
</ul>
</div>
<div id="step-2-create-docker-container-for-your-tool" class="section level3">
<h3 class="hasAnchor">
<a href="#step-2-create-docker-container-for-your-tool" class="anchor"></a>Step 2: Create Docker container for your tool</h3>
<p>Building a development environment is essential for developing your command line interface and your app. There are some principles:</p>
<ol style="list-style-type: decimal">
<li>First you need to check is there any existing container you can directly use, so you don’t have to make a new one. If you don’t know what to use, I will suggest you start with <code>rocker/hadleyverse</code>, it has lots of stuff like Markdown, knitr need and other Hadley’s popular packages.</li>
</ol>
<p>Official R Docker images is called “Rocker” project and is on <a href="https://github.com/rocker-org/rocker">GitHub</a>, please visit the page to find more details and Dockerfile.</p>
<table class="table">
<thead><tr class="header">
<th>Image</th>
<th align="center">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>rocker/r-base</code></td>
<td align="center">base package to build from</td>
</tr>
<tr class="even">
<td><code>rocker/r-devel</code></td>
<td align="center">base + R-devel from SVN</td>
</tr>
<tr class="odd">
<td><code>rocker/rstudio</code></td>
<td align="center">base + RStudio Server</td>
</tr>
<tr class="even">
<td><code>rocker/hadleyverse</code></td>
<td align="center">rstudio + Hadley’s packages, LaTeX</td>
</tr>
<tr class="odd">
<td><code>rocker/ropensci</code></td>
<td align="center">hadleyverse + rOpenSci packages</td>
</tr>
<tr class="even">
<td><code>rocker/r-devel-san</code></td>
<td align="center">base, SVN’s R-devel and SAN</td>
</tr>
</tbody>
</table>
<p>Bioconductor have a nice <a href="https://bioconductor.org/help/docker/">page</a> about the official Docker images, please read for more details.</p>
<table class="table">
<thead><tr class="header">
<th>Image (release branch)</th>
<th>Image (development branch)</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>bioconductor/release_base</code></td>
<td><code>bioconductor/devel_base</code></td>
</tr>
<tr class="even">
<td><code>bioconductor/release_core</code></td>
<td><code>bioconductor/devel_core</code></td>
</tr>
<tr class="odd">
<td><code>bioconductor/release_flow</code></td>
<td><code>bioconductor/devel_flow</code></td>
</tr>
<tr class="even">
<td><code>bioconductor/release_microarray</code></td>
<td><code>bioconductor/devel_microarray</code></td>
</tr>
<tr class="odd">
<td><code>bioconductor/release_proteomics</code></td>
<td><code>bioconductor/devel_proteomics</code></td>
</tr>
<tr class="even">
<td><code>bioconductor/release_sequencing</code></td>
<td><code>bioconductor/devel_sequencing</code></td>
</tr>
<tr class="odd">
<td><code>bioconductor/release_metabolomics</code></td>
<td><code>bioconductor/devel_metabolomics</code></td>
</tr>
</tbody>
</table>
<p>For example, you know there is an <code>runif</code> function in the <code>rocker/r-base</code> container, you can just do something like this. Please read another tutorial called “Describe and Execute CWL Tools/Workflows in R”, it introduced a simpler example with random number generator.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rbx &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Tool-class.html">Tool</a></span>(<span class="dt">id =</span> <span class="st">"runif"</span>,
            <span class="dt">label =</span> <span class="st">"runif"</span>,
            <span class="dt">hints =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(<span class="kw"><a href="../reference/requirements.html">docker</a></span>(<span class="dt">pull =</span> <span class="st">"rocker/r-base"</span>),
                                 <span class="kw"><a href="../reference/requirements.html">cpu</a></span>(<span class="dv">1</span>), <span class="kw"><a href="../reference/requirements.html">mem</a></span>(<span class="dv">2000</span>)),
            <span class="dt">baseCommand =</span> <span class="st">"Rscript -e 'runif(100)'"</span>,
            <span class="dt">stdout =</span> <span class="st">"output.txt"</span>,
            <span class="dt">outputs =</span> <span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"random"</span>, <span class="dt">glob =</span> <span class="st">"*.txt"</span>))
rbx<span class="op">$</span><span class="kw">toJSON</span>(<span class="dt">pretty =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>{
  "sbg:id": "runif",
  "id": "#runif",
  "inputs": [],
  "outputs": [
    {
      "type": ["null", "File"],
      "label": "",
      "description": "",
      "streamable": false,
      "default": "",
      "id": "#random",
      "outputBinding": {
        "glob": "*.txt"
      }
    }
  ],
  "requirements": [],
  "hints": [
    {
      "class": "DockerRequirement",
      "dockerPull": "rocker/r-base"
    },
    {
      "class": "sbg:CPURequirement",
      "value": 1
    },
    {
      "class": "sbg:MemRequirement",
      "value": 2000
    }
  ],
  "label": "runif",
  "class": "CommandLineTool",
  "baseCommand": [
    "Rscript -e 'runif(100)'"
  ],
  "arguments": [],
  "stdout": "output.txt"
} </code></pre>
<p>You can directly follow this <a href="http://docs.cancergenomicscloud.org/docs/the-tool-editor">tutorial</a>, paste your JSON into your tool editor by click “import”. Then “save” and “run”, you will be able to run your first application on CGC with no parameters and input files.</p>
<ol start="2" style="list-style-type: decimal">
<li>If you don’t want to make a new container with command line interface and you can simply insert script temporarily for existing container. You can do things as quick as this, by provide a script in the ‘content’ of ‘fileDef’.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># provide scripts</span>
<span class="co"># make a new script file</span>
fd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/requirements.html">fileDef</a></span>(<span class="dt">name =</span> <span class="st">"runif.R"</span>,
              <span class="dt">content =</span> <span class="st">"sed.seed(1)</span>
<span class="st">                   runif(100)"</span>)
rbx &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Tool-class.html">Tool</a></span>(<span class="dt">id =</span> <span class="st">"runif"</span>,
            <span class="dt">label =</span> <span class="st">"runif"</span>,
            <span class="dt">hints =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(<span class="kw"><a href="../reference/requirements.html">docker</a></span>(<span class="dt">pull =</span> <span class="st">"rocker/r-base"</span>),
                                 <span class="kw"><a href="../reference/requirements.html">cpu</a></span>(<span class="dv">1</span>), <span class="kw"><a href="../reference/requirements.html">mem</a></span>(<span class="dv">2000</span>)),
            <span class="dt">requirements =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(fd),
            <span class="dt">baseCommand =</span> <span class="st">"Rscript runif.R"</span>,  <span class="co"># run script you created.</span>
            <span class="dt">stdout =</span> <span class="st">"output.txt"</span>,
            <span class="dt">outputs =</span> <span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"random"</span>, <span class="dt">glob =</span> <span class="st">"*.txt"</span>))</code></pre></div>
<p><strong>Note</strong>: in the above example, I made a mistake on purpose, so try to debug on the platform if the task fails : )</p>
<p>I will introduce “Tool” function, in the later section, don’t worry.</p>
<ol start="3" style="list-style-type: decimal">
<li>For advanced developer/users: if you think a cool command line interface is worth doing, and convenient in any case, then go ahead and make one and create our own container will always be a better idea. this allow you to provide much formal interface at both command line level or container level.</li>
</ol>
<p>Here is Dockerfile I used to generate the workflow I need</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fl &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/system.file">system.file</a></span>(<span class="st">"docker/rnaseqGene"</span>, <span class="st">"Dockerfile"</span>,
                  <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)</code></pre></div>
<p>Here is the current content of the <code>Dockerfile</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">cat</span>(<span class="kw">readLines</span>(fl), <span class="dt">sep =</span> <span class="st">'</span><span class="ch">\n</span><span class="st">'</span>)</code></pre></div>
<pre><code>FROM rocker/tidyverse

MAINTAINER "Tengfei Yin" yintengfei@gmail.com

RUN Rscript -e 'source("http://bioconductor.org/workflows.R"); workflowInstall("rnaseqGene")'

ADD src/performDE.R /usr/local/bin/
RUN mkdir /report
ADD report/rnaseqGene.Rmd /report/

RUN chmod a+x /usr/local/bin/performDE.R \
    &amp;&amp; chmod -R a+x /report</code></pre>
<p>It does a couple of things:</p>
<ol style="list-style-type: decimal">
<li>Install workflow I need and all dependencies</li>
<li>Insert command line interface I created, make it executable and in <code>PATH</code>
</li>
<li>Insert full report template I am using</li>
</ol>
<p>In the next section, I will show you how to create command line interface.</p>
</div>
<div id="step-3-create-your-command-line-interface" class="section level3">
<h3 class="hasAnchor">
<a href="#step-3-create-your-command-line-interface" class="anchor"></a>Step 3: Create your command line interface</h3>
<p>In this step, I am going to:</p>
<ol style="list-style-type: decimal">
<li>Use R package <code>docopt</code> to make R command line interface</li>
<li>Insert my script into the container I made</li>
<li>Test function inside container</li>
</ol>
<p>Here is the command line I am using it’s called “performDE.R”</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fl &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/system.file">system.file</a></span>(<span class="st">"docker/rnaseqGene/src"</span>, <span class="st">"performDE.R"</span>,
                  <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)</code></pre></div>
<p>Here is the current content of the <code>Dockerfile</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">cat</span>(<span class="kw">readLines</span>(fl), <span class="dt">sep =</span> <span class="st">'</span><span class="ch">\n</span><span class="st">'</span>)</code></pre></div>
<pre><code>#!/usr/bin/Rscript
"usage: performDE.R [options]

options:
--bamfiles=&lt;file&gt; bamfiles
--design=&lt;file&gt; design data frame
--gtffile=&lt;file&gt; gene feature file
--format=&lt;string&gt; pdf or html. [default: html]
" -&gt; doc

library("docopt")
opts &lt;- docopt(doc)

.design &lt;- opts$design
if(is.null(.design))
    .design &lt;- system.file("extdata/sample_table.csv", package="airway")

.bamfiles &lt;- opts$bamfiles
if(is.null(.bamfiles)){
    .bamfiles &lt;- list.files(system.file("extdata", package="airway"), "*.bam",
                            full.names = TRUE)
}else{
    .bamfiles &lt;- strsplit(opts$bamfiles, ",")[[1]]
}

.gtffile &lt;- opts$gtffile
if(is.null(.gtffile))
    .gtffile &lt;- system.file("extdata/Homo_sapiens.GRCh37.75_subset.gtf",
                            package="airway")
## create param list
lst &lt;- list(design = normalizePath(.design),
            gtffile = normalizePath(.gtffile),
            bamfiles = normalizePath(.bamfiles),
            currentPath = normalizePath("."))


## render the report
.format &lt;- switch(opts$format,
                  "pdf" = "pdf_document",
                  "html" = "html_document",
                  {"pdf_document"})

rmarkdown::render("/report/rnaseqGene.Rmd", .format,
                  output_dir = ".",
                  params = lst)

# For rabix execution in the cloud,
# this is now a workaround to move intermediate files</code></pre>
<p>I am using <code>docopt</code> package instead of <code>commandArgs</code> because it allows you to create a formal interface. For example, you have <code>'--help'</code> from command line for free. Feel free to use other methods like <code>commandArgs()</code> in R.</p>
<pre><code>$./performDE.R --help
Loading required package: methods
usage: performDE.R [options]

options:
--bamfiles=&lt;file&gt; bamfiles
--design=&lt;file&gt; design data frame
--gtffile=&lt;file&gt; gene feature file
--format=&lt;string&gt; pdf or html. [default: html]</code></pre>
<p>Sometime you want to produce a report, so in the end you will notice how I pass parameters from command line to a report.</p>
</div>
<div id="step-4-add-default-report-template-to-your-app" class="section level3">
<h3 class="hasAnchor">
<a href="#step-4-add-default-report-template-to-your-app" class="anchor"></a>Step 4: Add default report template to your app</h3>
<p>As a developer, if you always have a fixed report template provided for you tool, you can hard-code your template into your Docker container, like you noticed in the Dockerfile I created, I insert a report R Markdown template. And in the command line interface, in the end, I pass it to <code><a href="http://www.rdocumentation.org/packages/rmarkdown/topics/render">rmarkdown::render</a></code> function as parameters ‘param’. In this way, you define your input in the header of R Markdown. Examples like this, this template I used here, is exactly the same report on Bioconductor’s RNA-Seq workflow <a href="https://www.bioconductor.org/help/workflows/rnaseqGene/">website</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fl &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/system.file">system.file</a></span>(<span class="st">"docker/rnaseqGene/report"</span>, <span class="st">"rnaseqGene.Rmd"</span>,
                  <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)</code></pre></div>
<p>Here is the current content (first 50 lines of the whole report) of the template:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">cat</span>(<span class="kw">readLines</span>(fl, <span class="dt">n =</span> <span class="dv">50</span>), <span class="dt">sep =</span> <span class="st">'</span><span class="ch">\n</span><span class="st">'</span>)</code></pre></div>
<pre><code>---
title: "Uniform random number generator example"
date: September 9, 2015
output:
  pdf_document:
    toc: true
    toc_depth: 2
params:
  bamfiles: ""
  design: ""
  gtffile: ""
  currentPath: "."
---

# Acknowledgement

This RNA-seq test report is generated based on Biocondcutor workflow,
for experimental purpose, I tweaked the contents. Instead of a test
data, I am createing command line interface based on this document
that take new data and generate this report. It's experiments about
docker, cwl and rabix.

&lt;script type="text/javascript"
  src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"&gt;
&lt;/script&gt;

__original documentaion comes from__

# RNA-seq workflow: gene-level exploratory analysis and differential expression

Michael Love [1], Simon Anders [2,3], Vladislav Kim [3], Wolfgang Huber [3]

[1] Department of Biostatistics, Dana-Farber Cancer Institute and
Harvard School of Public Health, Boston, US; 

[2] Institute for Molecular Medicine Finland (FIMM), Helsinki, Finland;

[3] European Molecular Biology Laboratory (EMBL), Heidelberg, Germany.

```{r style, echo=FALSE, message=FALSE, warning=FALSE, results="asis"}
library("knitr")
library("rmarkdown")
options(width = 100)
opts_knit$set(root.dir = params$currentPath)
opts_chunk$set(message = FALSE, error = FALSE,
               warning = FALSE, fig.width = 5, fig.height = 5)
```

# Contents</code></pre>
<p>See the header, you will see ‘params’ which is passed from <code><a href="http://www.rdocumentation.org/packages/rmarkdown/topics/render">rmarkdown::render</a></code>, so you can use it directly in the report like <code>params$bamfiles</code>.</p>
<p>Now you have</p>
<ul>
<li><code>Dockerfile</code></li>
<li>Command line interface with report template</li>
</ul>
<p>You are ready to build Docker container and push it to the registry. You can choose to use registry like dockerhub or use CGC’s own Docker registry (<code>cgc-images.sbgenomics.com</code>). To learn how to use CGC Image registry, please read out <a href="http://docs.cancergenomicscloud.org/docs/the-cgc-image-registry">tutorial</a>.</p>
</div>
<div id="step-5-describe-your-tool-in-r-into-cwl" class="section level3">
<h3 class="hasAnchor">
<a href="#step-5-describe-your-tool-in-r-into-cwl" class="anchor"></a>Step 5: Describe your tool in R into CWL</h3>
<p>Finally it’s time to describe your tool in R (CWL format)!</p>
<p>Well, you are always encouraged to use graphical user interface, but it is at the same time fun to learn how to do it in R, so you could script how you build it together, like what I did here, for every single tool JSON, I have an R script called <code>generator.R</code> in the same folder, so I can always trace back.</p>
<p>For example, you can see the same in the <code>sbg/sevenbridges-r</code> GitHub repo, under <a href="https://github.com/sbg/sevenbridges-r/tree/master/inst/docker">inst/docker</a>, you will see three examples, 1. package Docker 2. RNA-Seq tool 3. report tool, under each folder, you will see 1. one <code>Dockerfile</code>, 2. <code>src/</code> for command line 3. <code>report/</code> for report template and 4. rabix a generator file and a JSON.</p>
<p><img src="https://i.imgur.com/Wnf015S.png" width="100%" class="img-responsive" alt="rstudio-dir"></p>
<p><img src="https://i.imgur.com/OYoa6kq.png" width="100%" class="img-responsive" alt="rstudio-file"></p>
<p>Tool is the simple basic unit of a workflow, you can put the whole flow in one container and one tool, it of course works, just make it hard to factorize components. This is the exact example, I can make one tool for <code>DESeq2</code> and one tool for <code>Rsamtools</code>, I can also put everything I need in one tool and provide single functionality.</p>
<p>Note: you can use single Docker image, but describe as many tools as you want if it contains what you need such as different command.</p>
<p>Follow the example to create Tool with <code>Tool</code> function. It’s straightforward. Especially if you are familiar with Seven Bridges Tool editor already.</p>
<p>Hints: please pay attention to how I create</p>
<ul>
<li>File list: via ItemArray(“File”) or “File…” this allow you to input multiple files form the task page.</li>
<li>Single File: just “File”, only single file allowed.</li>
<li>Expression to specify the javascript expression (note:it’s convenient to do it with graphic user interface, because you can directly see the result of the expression)</li>
<li>Enum: call ‘enum’.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rbx &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Tool-class.html">Tool</a></span>(<span class="dt">id =</span> <span class="st">"rnaseqGene"</span>,
            <span class="dt">label =</span> <span class="st">"rnaseqgene"</span>,
            <span class="dt">description =</span> <span class="st">"A RNA-seq Differiencial Expression Flow and Report"</span>,
            <span class="dt">hints =</span> <span class="kw"><a href="../reference/requirements.html">requirements</a></span>(<span class="kw"><a href="../reference/requirements.html">docker</a></span>(<span class="dt">pull =</span> <span class="st">"tengfei/rnaseqgene"</span>), <span class="kw"><a href="../reference/requirements.html">cpu</a></span>(<span class="dv">1</span>), <span class="kw"><a href="../reference/requirements.html">mem</a></span>(<span class="dv">2000</span>)),
            <span class="dt">baseCommand =</span> <span class="st">"performDE.R"</span>,
            <span class="dt">inputs =</span> <span class="kw">list</span>(
                <span class="kw"><a href="../reference/shorthand-cwl.html">input</a></span>(
                    <span class="dt">id =</span> <span class="st">"bamfiles"</span>, <span class="dt">label =</span> <span class="st">"bam files"</span>,
                    <span class="dt">description =</span> <span class="st">"a list of bam files"</span>,
                    <span class="dt">type =</span> <span class="st">"File..."</span>,  ## or type = ItemArray("File")
                    <span class="dt">prefix =</span> <span class="st">"--bamfiles"</span>,
                    <span class="dt">itemSeparator =</span> <span class="st">","</span>
                ),
                <span class="kw"><a href="../reference/shorthand-cwl.html">input</a></span>(
                    <span class="dt">id =</span> <span class="st">"design"</span>, <span class="dt">label =</span> <span class="st">"design matrix"</span>,
                    <span class="dt">type =</span> <span class="st">"File"</span>,
                    <span class="dt">prefix =</span> <span class="st">"--design"</span>
                ),
                <span class="kw"><a href="../reference/shorthand-cwl.html">input</a></span>(
                    <span class="dt">id =</span> <span class="st">"gtffile"</span>, <span class="dt">label =</span>  <span class="st">"gene feature files"</span>,
                    <span class="dt">type =</span> <span class="st">"File"</span>,
                    <span class="dt">prefix =</span> <span class="st">"--gtffile"</span>
                ),
                <span class="kw"><a href="../reference/shorthand-cwl.html">input</a></span>(
                    <span class="dt">id =</span> <span class="st">"format"</span>, <span class="dt">label =</span>  <span class="st">"report foramt html or pdf"</span>,
                    <span class="dt">type =</span> <span class="kw"><a href="../reference/Enum.html">enum</a></span>(<span class="st">"format"</span>, <span class="kw">c</span>(<span class="st">"pdf"</span>, <span class="st">"html"</span>)),
                    <span class="dt">prefix =</span> <span class="st">"--format"</span>
                )
            ),
            <span class="dt">outputs =</span> <span class="kw">list</span>(
                <span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"report"</span>, <span class="dt">label =</span> <span class="st">"report"</span>,
                       <span class="dt">description =</span> <span class="st">"A reproducible report created by Rmarkdown"</span>,
                       <span class="dt">glob =</span> <span class="kw"><a href="../reference/Expression-class.html">Expression</a></span>(<span class="dt">engine =</span> <span class="st">"#cwl-js-engine"</span>,
                                         <span class="dt">script =</span> <span class="st">"x = $job[['inputs']][['format']];</span>
<span class="st">                                                  if(x == 'undefined' || x == null){</span>
<span class="st">                                                   x = 'html';</span>
<span class="st">                                                    };</span>
<span class="st">                                                  'rnaseqGene.' +  x"</span>)),
                <span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"heatmap"</span>, <span class="dt">label =</span> <span class="st">"heatmap"</span>,
                       <span class="dt">description =</span> <span class="st">"A heatmap plot to show the Euclidean distance between samples"</span>,
                       <span class="dt">glob =</span> <span class="st">"heatmap.pdf"</span>),
                <span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"count"</span>, <span class="dt">label =</span> <span class="st">"count"</span>,
                       <span class="dt">description =</span> <span class="st">"Reads counts matrix"</span>,
                       <span class="dt">glob =</span> <span class="st">"count.csv"</span>),
                <span class="kw"><a href="../reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"de"</span>, <span class="dt">label =</span> <span class="st">"Differential expression table"</span>,
                       <span class="dt">description =</span> <span class="st">"Differential expression table"</span>,
                       <span class="dt">glob =</span> <span class="st">"de.csv"</span>)
                ))</code></pre></div>
<p>By default it output YAML, but you can print it into JSON as well.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rbx</code></pre></div>
<pre><code>sbg:id: rnaseqGene
id: '#rnaseqGene'
inputs:
- type:
  - 'null'
  - items: File
    type: array
  label: bam files
  description: a list of bam files
  streamable: no
  default: ''
  id: '#bamfiles'
  inputBinding:
    position: 0
    prefix: --bamfiles
    separate: yes
    itemSeparator: ','
    sbg:cmdInclude: yes
  required: no
- type:
  - 'null'
  - File
  label: design matrix
  description: ''
  streamable: no
  default: ''
  id: '#design'
  inputBinding:
    position: 0
    prefix: --design
    separate: yes
    sbg:cmdInclude: yes
  required: no
- type:
  - 'null'
  - File
  label: gene feature files
  description: ''
  streamable: no
  default: ''
  id: '#gtffile'
  inputBinding:
    position: 0
    prefix: --gtffile
    separate: yes
    sbg:cmdInclude: yes
  required: no
- type:
  - 'null'
  - name: format
    symbols:
    - pdf
    - html
    type: enum
  label: report foramt html or pdf
  description: ''
  streamable: no
  default: ''
  id: '#format'
  inputBinding:
    position: 0
    prefix: --format
    separate: yes
    sbg:cmdInclude: yes
  required: no
outputs:
- type:
  - 'null'
  - File
  label: report
  description: A reproducible report created by Rmarkdown
  streamable: no
  default: ''
  id: '#report'
  outputBinding:
    glob:
      engine: '#cwl-js-engine'
      script: |-
        x = $job[['inputs']][['format']];
                                                          if(x == 'undefined' || x == null){
                                                           x = 'html';
                                                            };
                                                          'rnaseqGene.' +  x
      class: Expression
- type:
  - 'null'
  - File
  label: heatmap
  description: A heatmap plot to show the Euclidean distance between samples
  streamable: no
  default: ''
  id: '#heatmap'
  outputBinding:
    glob: heatmap.pdf
- type:
  - 'null'
  - File
  label: count
  description: Reads counts matrix
  streamable: no
  default: ''
  id: '#count'
  outputBinding:
    glob: count.csv
- type:
  - 'null'
  - File
  label: Differential expression table
  description: Differential expression table
  streamable: no
  default: ''
  id: '#de'
  outputBinding:
    glob: de.csv
requirements: []
hints:
- class: DockerRequirement
  dockerPull: tengfei/rnaseqgene
- class: sbg:CPURequirement
  value: 1
- class: sbg:MemRequirement
  value: 2000
label: rnaseqgene
description: A RNA-seq Differiencial Expression Flow and Report
class: CommandLineTool
baseCommand:
- performDE.R
arguments: []</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rbx<span class="op">$</span><span class="kw">toJSON</span>(<span class="dt">pretty =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>{
  "sbg:id": "rnaseqGene",
  "id": "#rnaseqGene",
  "inputs": [
    {
      "type": [
        "null",
        {
          "items": "File",
          "type": "array"
        }
      ],
      "label": "bam files",
      "description": "a list of bam files",
      "streamable": false,
      "default": "",
      "id": "#bamfiles",
      "inputBinding": {
        "position": 0,
        "prefix": "--bamfiles",
        "separate": true,
        "itemSeparator": ",",
        "sbg:cmdInclude": true
      },
      "required": false
    },
    {
      "type": ["null", "File"],
      "label": "design matrix",
      "description": "",
      "streamable": false,
      "default": "",
      "id": "#design",
      "inputBinding": {
        "position": 0,
        "prefix": "--design",
        "separate": true,
        "sbg:cmdInclude": true
      },
      "required": false
    },
    {
      "type": ["null", "File"],
      "label": "gene feature files",
      "description": "",
      "streamable": false,
      "default": "",
      "id": "#gtffile",
      "inputBinding": {
        "position": 0,
        "prefix": "--gtffile",
        "separate": true,
        "sbg:cmdInclude": true
      },
      "required": false
    },
    {
      "type": [
        "null",
        {
          "name": "format",
          "symbols": ["pdf", "html"],
          "type": "enum"
        }
      ],
      "label": "report foramt html or pdf",
      "description": "",
      "streamable": false,
      "default": "",
      "id": "#format",
      "inputBinding": {
        "position": 0,
        "prefix": "--format",
        "separate": true,
        "sbg:cmdInclude": true
      },
      "required": false
    }
  ],
  "outputs": [
    {
      "type": ["null", "File"],
      "label": "report",
      "description": "A reproducible report created by Rmarkdown",
      "streamable": false,
      "default": "",
      "id": "#report",
      "outputBinding": {
        "glob": {
          "engine": "#cwl-js-engine",
          "script": "x = $job[['inputs']][['format']];\n                                                  if(x == 'undefined' || x == null){\n                                                   x = 'html';\n                                                    };\n                                                  'rnaseqGene.' +  x",
          "class": "Expression"
        }
      }
    },
    {
      "type": ["null", "File"],
      "label": "heatmap",
      "description": "A heatmap plot to show the Euclidean distance between samples",
      "streamable": false,
      "default": "",
      "id": "#heatmap",
      "outputBinding": {
        "glob": "heatmap.pdf"
      }
    },
    {
      "type": ["null", "File"],
      "label": "count",
      "description": "Reads counts matrix",
      "streamable": false,
      "default": "",
      "id": "#count",
      "outputBinding": {
        "glob": "count.csv"
      }
    },
    {
      "type": ["null", "File"],
      "label": "Differential expression table",
      "description": "Differential expression table",
      "streamable": false,
      "default": "",
      "id": "#de",
      "outputBinding": {
        "glob": "de.csv"
      }
    }
  ],
  "requirements": [],
  "hints": [
    {
      "class": "DockerRequirement",
      "dockerPull": "tengfei/rnaseqgene"
    },
    {
      "class": "sbg:CPURequirement",
      "value": 1
    },
    {
      "class": "sbg:MemRequirement",
      "value": 2000
    }
  ],
  "label": "rnaseqgene",
  "description": "A RNA-seq Differiencial Expression Flow and Report",
  "class": "CommandLineTool",
  "baseCommand": [
    "performDE.R"
  ],
  "arguments": []
} </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rbx<span class="op">$</span><span class="kw">toJSON</span>()</code></pre></div>
<pre><code>{"sbg:id":"rnaseqGene","id":"#rnaseqGene","inputs":[{"type":["null",{"items":"File","type":"array"}],"label":"bam files","description":"a list of bam files","streamable":false,"default":"","id":"#bamfiles","inputBinding":{"position":0,"prefix":"--bamfiles","separate":true,"itemSeparator":",","sbg:cmdInclude":true},"required":false},{"type":["null","File"],"label":"design matrix","description":"","streamable":false,"default":"","id":"#design","inputBinding":{"position":0,"prefix":"--design","separate":true,"sbg:cmdInclude":true},"required":false},{"type":["null","File"],"label":"gene feature files","description":"","streamable":false,"default":"","id":"#gtffile","inputBinding":{"position":0,"prefix":"--gtffile","separate":true,"sbg:cmdInclude":true},"required":false},{"type":["null",{"name":"format","symbols":["pdf","html"],"type":"enum"}],"label":"report foramt html or pdf","description":"","streamable":false,"default":"","id":"#format","inputBinding":{"position":0,"prefix":"--format","separate":true,"sbg:cmdInclude":true},"required":false}],"outputs":[{"type":["null","File"],"label":"report","description":"A reproducible report created by Rmarkdown","streamable":false,"default":"","id":"#report","outputBinding":{"glob":{"engine":"#cwl-js-engine","script":"x = $job[['inputs']][['format']];\n                                                  if(x == 'undefined' || x == null){\n                                                   x = 'html';\n                                                    };\n                                                  'rnaseqGene.' +  x","class":"Expression"}}},{"type":["null","File"],"label":"heatmap","description":"A heatmap plot to show the Euclidean distance between samples","streamable":false,"default":"","id":"#heatmap","outputBinding":{"glob":"heatmap.pdf"}},{"type":["null","File"],"label":"count","description":"Reads counts matrix","streamable":false,"default":"","id":"#count","outputBinding":{"glob":"count.csv"}},{"type":["null","File"],"label":"Differential expression table","description":"Differential expression table","streamable":false,"default":"","id":"#de","outputBinding":{"glob":"de.csv"}}],"requirements":[],"hints":[{"class":"DockerRequirement","dockerPull":"tengfei/rnaseqgene"},{"class":"sbg:CPURequirement","value":1},{"class":"sbg:MemRequirement","value":2000}],"label":"rnaseqgene","description":"A RNA-seq Differiencial Expression Flow and Report","class":"CommandLineTool","baseCommand":["performDE.R"],"arguments":[]} </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># or write to external file</span>
<span class="co"># fl &lt;- "~/Downloads/rnaseqGene.json"</span>
<span class="co"># write(rbx$toJSON(pretty = TRUE), fl)</span></code></pre></div>
<p>Now you want to add app to your project <code>p</code>, by calling <code>app_add</code> method, the first argument is name, the second is either a CWL JSON file, <code>Tool</code> object, or <code>Workflow</code> object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># add App you just created</span>
(rna.app &lt;-<span class="st"> </span>p<span class="op">$</span><span class="kw">app_add</span>(<span class="st">"rnaseqgene"</span>, rbx))</code></pre></div>
<p>Please go check your app in your project, check input output and how it maps to the UI.</p>
</div>
<div id="step-6-execute-your-tool-with-a-new-task-via-r-api" class="section level3">
<h3 class="hasAnchor">
<a href="#step-6-execute-your-tool-with-a-new-task-via-r-api" class="anchor"></a>Step 6: Execute your tool with a new task via R API</h3>
<p>Now let’s create a task and execute it with example files. You need to pass unique file id as input, so the first thing is to get file id you need for that project.</p>
<p>Now let’s import some files to your project for this tutorial, you can also do it via our API client by call <code>upload</code> function on project object. It supports</p>
<ul>
<li>Single file, multiple file, and folder (recursively all files) as first arugment</li>
<li>
<code>name</code>: a new name (with single file upload)</li>
<li>
<code>overwrite = TRUE</code> to overwrite existing file</li>
<li>
<code>metadata</code>: a list of meta otherwise search for the same file name ended with “.meta”</li>
</ul>
<p>For example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fl &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"sample1.fastq"</span>, <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
(p &lt;-<span class="st"> </span>a<span class="op">$</span><span class="kw">project</span>(<span class="dt">id =</span> <span class="st">"tengfei/quickstart"</span>))
<span class="co"># by default load .meta for the file</span>
p<span class="op">$</span><span class="kw">upload</span>(fl, <span class="dt">overwrite =</span> <span class="ot">TRUE</span>)
<span class="co"># pass metadata</span>
p<span class="op">$</span><span class="kw">upload</span>(fl, <span class="dt">overwrite =</span> <span class="ot">TRUE</span>, <span class="dt">metadata =</span> <span class="kw">list</span>(<span class="dt">library_id =</span> <span class="st">"testid2"</span>, <span class="dt">platform =</span> <span class="st">"Illumina x11"</span>))
<span class="co"># rename</span>
p<span class="op">$</span><span class="kw">upload</span>(fl, <span class="dt">overwrite =</span> <span class="ot">TRUE</span>, <span class="dt">name =</span> <span class="st">"sample_new_name.fastq"</span>,
         <span class="dt">metadata =</span> <span class="kw">list</span>(<span class="dt">library_id =</span> <span class="st">"new_id"</span>))

<span class="co"># upload folder</span>
dir.ext &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
p<span class="op">$</span><span class="kw">upload</span>(dir.ext, <span class="dt">overwrite =</span> <span class="ot">TRUE</span>)

<span class="co"># upload file list</span>
fls &lt;-<span class="st"> </span><span class="kw">list.files</span>(dir.ext, <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)
p<span class="op">$</span><span class="kw">upload</span>(fls, <span class="dt">overwrite =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>For now try using our graphic user interface to import all files listed here:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">download.fl &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/system.file">system.file</a></span>(<span class="st">"extdata/download.txt"</span>, <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
<span class="kw">cat</span>(<span class="kw">readLines</span>(download.fl), <span class="dt">sep =</span> <span class="st">'</span><span class="ch">\n</span><span class="st">'</span>)</code></pre></div>
<pre><code>Warning in readLines(download.fl): incomplete final line found on '/usr/
local/lib/R/3.4/site-library/sevenbridges/extdata/download.txt'</code></pre>
<pre><code>https://raw.githubusercontent.com/tengfei/resource/master/2016/04-01-hackathon/data/example_report.Rmd
https://raw.githubusercontent.com/tengfei/resource/master/2016/04-01-hackathon/data/hello-markdown.Rmd
https://raw.githubusercontent.com/tengfei/resource/master/2016/04-01-hackathon/data/hello.tar
https://raw.githubusercontent.com/tengfei/resource/master/2016/04-01-hackathon/data/Homo_sapiens.GRCh37.75_subset.gtf
https://raw.githubusercontent.com/tengfei/resource/master/2016/04-01-hackathon/data/sample_table.csv
https://raw.githubusercontent.com/tengfei/resource/master/2016/04-01-hackathon/data/scatter_plot.tar
https://raw.githubusercontent.com/tengfei/resource/master/2016/04-01-hackathon/data/SRR1039508_subset.bam
https://raw.githubusercontent.com/tengfei/resource/master/2016/04-01-hackathon/data/SRR1039509_subset.bam
https://raw.githubusercontent.com/tengfei/resource/master/2016/04-01-hackathon/data/SRR1039512_subset.bam
https://raw.githubusercontent.com/tengfei/resource/master/2016/04-01-hackathon/data/SRR1039513_subset.bam
https://raw.githubusercontent.com/tengfei/resource/master/2016/04-01-hackathon/data/SRR1039516_subset.bam
https://raw.githubusercontent.com/tengfei/resource/master/2016/04-01-hackathon/data/SRR1039517_subset.bam
https://raw.githubusercontent.com/tengfei/resource/master/2016/04-01-hackathon/data/SRR1039520_subset.bam
https://raw.githubusercontent.com/tengfei/resource/master/2016/04-01-hackathon/data/SRR1039521_subset.bam</code></pre>
<p>To use the API to uplaod, let’s download it to a folder and upload via API.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">td &lt;-<span class="st"> </span><span class="kw">tempfile</span>()
<span class="kw">dir.create</span>(td)
<span class="cf">for</span>(f <span class="cf">in</span> <span class="kw">readLines</span>(download.fl)){
    <span class="kw">download.file</span>(f, <span class="kw">file.path</span>(td, <span class="kw">basename</span>(f)))
}
<span class="co"># double check</span>
<span class="kw">list.files</span>(td)
<span class="co"># upload to the project you created</span>
p<span class="op">$</span><span class="kw">upload</span>(td)</code></pre></div>
<p>Copy this list, and then in your project, click “add files”, and choose “import from ftp” (<a href="http://docs.cancergenomicscloud.org/docs/copy-files-to-a-project">tutorial</a>).</p>
<p>When it’s finished, refresh your file page, you will be able to see all of them. The cool thing is that you can search file by ‘name’ not by id, it support fuzzy pattern match.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get file id you need as inout</span>
(bamfiles.in &lt;-<span class="st"> </span>p<span class="op">$</span><span class="kw">file</span>(<span class="st">".bam"</span>))
(design.in &lt;-<span class="st"> </span>p<span class="op">$</span><span class="kw">file</span>(<span class="st">"sample_table.csv"</span>))
(gtf.in &lt;-<span class="st"> </span>p<span class="op">$</span><span class="kw">file</span>(<span class="st">"Homo_sapiens.GRCh37.75_subset.gtf"</span>))</code></pre></div>
<p>Note: you can also passed a list of files like this, to give exact file names</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bam1 &lt;-<span class="st"> </span>p<span class="op">$</span><span class="kw">file</span>(<span class="st">"SRR1039516_subset.bam"</span>)
bam2 &lt;-<span class="st"> </span>p<span class="op">$</span><span class="kw">file</span>(<span class="st">"SRR1039512_subset.bam"</span>)
bamfiles.in2 &lt;-<span class="st"> </span><span class="kw">list</span>(bam1, bam2)</code></pre></div>
<p>Now create a new draft task in your project, don’t forget to pass input.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># add a new Task</span>
(tsk &lt;-<span class="st"> </span>p<span class="op">$</span><span class="kw">task_add</span>(<span class="dt">name =</span> <span class="st">"RNA DE report new"</span>,
                   <span class="dt">description =</span> <span class="st">"RNA DE analysis report"</span>,
                   <span class="dt">app =</span> rna.app<span class="op">$</span>id,
                   <span class="dt">inputs =</span> <span class="kw">list</span>(<span class="dt">bamfiles =</span> bamfiles.in,
                                 <span class="dt">design =</span> design.in,
                                 <span class="dt">gtffile =</span> gtf.in)))

<span class="co"># don't forget to run a draft task</span>
tsk<span class="op">$</span><span class="kw">run</span>()</code></pre></div>
<p>To monitor the task, run following command, it will tell you when it’s finished, but this is not running in the background now.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># # monitor the task</span>
<span class="co"># tsk$monitor()</span></code></pre></div>
<p>A better way is to use the Task hook function, it’s flexible, you can hook any function to a task status. For example, when it’s complete download the files. Now try to send your self a text message : )</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/TaskHook.html">setTaskHook</a></span>(<span class="st">"completed"</span>, <span class="cf">function</span>() {
    tsk<span class="op">$</span><span class="kw"><a href="../reference/download-methods.html">download</a></span>(<span class="st">"~/Downloads"</span>)
})
tsk<span class="op">$</span><span class="kw">monitor</span>()</code></pre></div>
<p>To download all files from a completed tasks</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tsk<span class="op">$</span><span class="kw"><a href="../reference/download-methods.html">download</a></span>(<span class="st">"~/Downloads"</span>)</code></pre></div>
<p>To run task in batch mode, check <code><a href="../reference/batch.html">?batch</a></code> for more details. Here is an mock running:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># batch by items</span>
(tsk &lt;-<span class="st"> </span>p<span class="op">$</span><span class="kw">task_add</span>(<span class="dt">name =</span> <span class="st">"RNA DE report new batch 2"</span>,
                   <span class="dt">description =</span> <span class="st">"RNA DE analysis report"</span>,
                   <span class="dt">app =</span> rna.app<span class="op">$</span>id,
                   <span class="dt">batch =</span> <span class="kw"><a href="../reference/batch.html">batch</a></span>(<span class="dt">input =</span> <span class="st">"bamfiles"</span>),
                   <span class="dt">inputs =</span> <span class="kw">list</span>(<span class="dt">bamfiles =</span> bamfiles.in,
                                 <span class="dt">design =</span> design.in,
                                 <span class="dt">gtffile =</span> gtf.in)))

<span class="co"># batch by metadata, input files has to have metadata fields specified</span>
(tsk &lt;-<span class="st"> </span>p<span class="op">$</span><span class="kw">task_add</span>(<span class="dt">name =</span> <span class="st">"RNA DE report new batch 3"</span>,
                   <span class="dt">description =</span> <span class="st">"RNA DE analysis report"</span>,
                   <span class="dt">app =</span> rna.app<span class="op">$</span>id,
                   <span class="dt">batch =</span> <span class="kw"><a href="../reference/batch.html">batch</a></span>(<span class="dt">input =</span> <span class="st">"fastq"</span>,
                                 <span class="kw">c</span>(<span class="st">"metadata.sample_id"</span>, <span class="st">"metadata.library_id"</span>)),
                   <span class="dt">inputs =</span> <span class="kw">list</span>(<span class="dt">bamfiles =</span> bamfiles.in,
                                 <span class="dt">design =</span> design.in,
                                 <span class="dt">gtffile =</span> gtf.in)))</code></pre></div>
<p>For more details, check R API tutorial.</p>
</div>
</div>
<div id="build-a-reporting-tool" class="section level2">
<h2 class="hasAnchor">
<a href="#build-a-reporting-tool" class="anchor"></a>Build a reporting Tool</h2>
<div id="what-is-report-tool" class="section level3">
<h3 class="hasAnchor">
<a href="#what-is-report-tool" class="anchor"></a>What is report tool?</h3>
<p>In our last section, we demonstrate how to hardcode a R report template into your Docker container and your command line will output a new report with new input files.</p>
<p>But how about this, instead of coding template into Docker, I want to</p>
<ul>
<li>Provide multiple R markdown template as input</li>
<li>Provide multiple shiny template as input</li>
<li>Provide multiple liftr template as input</li>
<li>Collect files from a flow used for my report</li>
<li>Collect report from different tools of a single workflow and compile a summary report</li>
</ul>
<p>Here I am developing a report tool to support all these in a single tool and it uses two different engines to provide isolated environment to generate report, because each report has different dependencies.</p>
<ul>
<li>
<code>packrat</code> engine: isolated libraries</li>
<li>liftr engine (in progress): Docker in Docker</li>
</ul>
<p><img src="https://i.imgur.com/yPTNpqp.png" width="100%" class="img-responsive" alt="report-tool"></p>
<p><strong>Requirement for your report template</strong></p>
<p>This is how it works</p>
<ol style="list-style-type: decimal">
<li>An app folder is created for each template</li>
<li>In this root folder, you have <code>data</code>, <code>src</code>, <code>www</code> folders, on the interface, you can connect output into those folders, so data flows into it.</li>
<li>First copy what’s in your template, then copy those task output into those folders.</li>
<li>Now create isolated environment by <code>packrat</code> or <code>liftr</code>.</li>
<li>Compile your template into report or new Shiny app</li>
<li>If you pass shinyapp.io token, name, secret, will deploy shiny app for you!</li>
<li>So in your template, you have to know file name and location relative to your app root, this is the most important requirement now.</li>
</ol>
<p>If you are interested, you can still read my <code>Dockerfile</code>, command line, tool generator and JSON.</p>
<p><em>I will suggest you directly copy the json into your project.</em> Just to try add an app in a different way.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fl &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/system.file">system.file</a></span>(<span class="st">"docker/reporttool/rabix/reporttool.json"</span>,
                  <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
<span class="kw">cat</span>(<span class="kw">readLines</span>(fl), <span class="dt">sep=</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</code></pre></div>
<pre><code>{
  "sbg:id": "reporttool",
  "id": "#reporttool",
  "inputs": [
    {
      "type": [
        "null",
        {
          "items": "File",
          "type": "array"
        }
      ],
      "label": "",
      "description": "Shinay app template as zipped(.zip) or tar(tar.gz) file.",
      "streamable": false,
      "default": "",
      "id": "#shinyTemplate",
      "inputBinding": {
        "position": 0,
        "prefix": "--shinyTemplate",
        "separate": true,
        "itemSeparator": ",",
        "sbg:cmdInclude": true
      }
    },
    {
      "type": [
        "null",
        {
          "items": "File",
          "type": "array"
        }
      ],
      "label": "",
      "description": "Rmarkdown file template will be rendered by knitr",
      "streamable": false,
      "default": "",
      "id": "#knitrTemplate",
      "inputBinding": {
        "position": 0,
        "prefix": "--knitrTemplate",
        "separate": true,
        "itemSeparator": ",",
        "sbg:cmdInclude": true
      }
    },
    {
      "type": [
        "null",
        {
          "items": "File",
          "type": "array"
        }
      ],
      "label": "",
      "description": "Files to be included in data folder of app",
      "streamable": false,
      "default": "",
      "id": "#data",
      "inputBinding": {
        "position": 0,
        "prefix": "--data",
        "separate": true,
        "itemSeparator": ",",
        "sbg:cmdInclude": true
      }
    },
    {
      "type": [
        "null",
        {
          "items": "File",
          "type": "array"
        }
      ],
      "label": "",
      "description": "Files to be included in www folder of app",
      "streamable": false,
      "default": "",
      "id": "#www",
      "inputBinding": {
        "position": 0,
        "prefix": "--www",
        "separate": true,
        "itemSeparator": ",",
        "sbg:cmdInclude": true
      }
    },
    {
      "type": [
        "null",
        {
          "items": "File",
          "type": "array"
        }
      ],
      "label": "",
      "description": "Files to be included in src folder of app",
      "streamable": false,
      "default": "",
      "id": "#src",
      "inputBinding": {
        "position": 0,
        "prefix": "--src",
        "separate": true,
        "sbg:cmdInclude": true
      }
    },
    {
      "type": [
        "null",
        {
          "items": "File",
          "type": "array"
        }
      ],
      "label": "",
      "description": "Files to be included in root of app folder",
      "streamable": false,
      "default": "",
      "id": "#appFiles",
      "inputBinding": {
        "position": 0,
        "prefix": "--appFiles",
        "separate": true,
        "itemSeparator": ",",
        "sbg:cmdInclude": true
      }
    },
    {
      "type": ["null", "string"],
      "label": "",
      "description": "Name of account to save or remove, check shinyapps::setAccountInfo",
      "streamable": false,
      "default": "",
      "id": "#name",
      "inputBinding": {
        "position": 0,
        "prefix": "--name",
        "separate": true,
        "sbg:cmdInclude": true
      }
    },
    {
      "type": ["null", "string"],
      "label": "",
      "description": "User token for the account, check shinyapps::setAccountInfo",
      "streamable": false,
      "default": "",
      "id": "#token",
      "inputBinding": {
        "position": 0,
        "prefix": "--token",
        "separate": true,
        "sbg:cmdInclude": true
      }
    },
    {
      "type": ["null", "string"],
      "label": "",
      "description": "User secret for the account, check shinyapps::setAccountInfo",
      "streamable": false,
      "default": "",
      "id": "#secret",
      "inputBinding": {
        "position": 0,
        "prefix": "--secret",
        "separate": true,
        "sbg:cmdInclude": true
      }
    },
    {
      "type": ["null", "string"],
      "label": "",
      "description": "Optional; the kind of content being deployed (e.g. 'plot', 'document', or 'application').",
      "streamable": false,
      "default": "",
      "id": "#contentCategory",
      "inputBinding": {
        "position": 0,
        "prefix": "--contentCategory",
        "separate": true,
        "sbg:cmdInclude": true
      }
    },
    {
      "type": ["null", "string"],
      "label": "",
      "description": "Account to deploy application to. This parameter is only required for the initial deployment of an application when there are multiple accounts configured on the system (see accounts).",
      "streamable": false,
      "default": "",
      "id": "#account",
      "inputBinding": {
        "position": 0,
        "prefix": "--account",
        "separate": true,
        "sbg:cmdInclude": true
      }
    },
    {
      "type": ["null", "string"],
      "label": "",
      "description": "Server name. Required only if you use the same account name on multiple servers.",
      "streamable": false,
      "default": "",
      "id": "#server",
      "inputBinding": {
        "position": 0,
        "prefix": "--server",
        "separate": true,
        "sbg:cmdInclude": true
      }
    },
    {
      "type": ["null", "boolean"],
      "label": "",
      "description": "Request that no status information be printed to the console during the deployment.",
      "streamable": false,
      "default": "",
      "id": "#quiet",
      "inputBinding": {
        "position": 0,
        "prefix": "--quiet",
        "separate": true,
        "sbg:cmdInclude": true
      }
    },
    {
      "type": [
        "null",
        {
          "name": "engine",
          "symbols": ["packrat", "liftr"],
          "type": "enum"
        }
      ],
      "label": "",
      "description": "packrat or liftr (docker in docker) or NA [default: packrat]",
      "streamable": false,
      "default": "",
      "id": "#engine",
      "inputBinding": {
        "position": 0,
        "prefix": "--engine",
        "separate": true,
        "sbg:cmdInclude": true
      }
    }
  ],
  "outputs": [
    {
      "type": [
        "null",
        {
          "items": "File",
          "type": "array"
        }
      ],
      "label": "",
      "description": "compressed shiny app folder",
      "streamable": false,
      "default": "",
      "id": "#shinyapp",
      "outputBinding": {
        "glob": "*.tar"
      }
    },
    {
      "type": [
        "null",
        {
          "items": "File",
          "type": "array"
        }
      ],
      "label": "",
      "description": "report rendered as html from knitr template",
      "streamable": false,
      "default": "",
      "id": "#html_report",
      "outputBinding": {
        "glob": "*.html"
      }
    },
    {
      "type": [
        "null",
        {
          "items": "File",
          "type": "array"
        }
      ],
      "label": "",
      "description": "report rendered as pdf from knitr template",
      "streamable": false,
      "default": "",
      "id": "#pdf_report",
      "outputBinding": {
        "glob": "*.pdf"
      }
    }
  ],
  "requirements": [],
  "hints": [
    {
      "class": "DockerRequirement",
      "dockerPull": "tengfei/reporttool",
      "dockerLoad": "",
      "dockerFile": "",
      "dockerImageId": "",
      "dockerOutputDirectory": ""
    }
  ],
  "label": "reporttool",
  "description": "Reporiting tools support you pass shiny app and knitr Rmakrdown template",
  "class": "CommandLineTool",
  "baseCommand": [
    "report.R"
  ],
  "arguments": [],
  "context": ""
}</code></pre>
<p>Or just use API to add the raw JSON file</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># directly add json file</span>
p &lt;-<span class="st"> </span>a<span class="op">$</span><span class="kw">project</span>(<span class="dt">id =</span> <span class="st">"tengfei/hackathon"</span>)
(report.app &lt;-<span class="st"> </span>p<span class="op">$</span><span class="kw">app_add</span>(<span class="st">"report-tool"</span>, fl))</code></pre></div>
<p>Checkout the Dockerfile</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fl &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/system.file">system.file</a></span>(<span class="st">"docker/reporttool/Dockerfile"</span>,
                  <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)
<span class="kw">cat</span>(<span class="kw">readLines</span>(fl), <span class="dt">sep=</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</code></pre></div>
<pre><code>FROM rocker/tidyverse

MAINTAINER "Tengfei Yin" yintengfei@gmail.com

RUN rm -f /var/lib/dpkg/available \
    &amp;&amp; rm -rf  /var/cache/apt/* \
    &amp;&amp; apt-get update \
    &amp;&amp; apt-get install -y libssl-dev

RUN R -e "install.packages(c('packrat', 'devtools', 'rsconnect', 'shiny', 'rmarkdown'), repos = 'https://cran.rstudio.com/')"

RUN R -e "devtools::install_github('road2stat/liftr', ref = 'devel')"

ADD src/report.R /usr/local/bin/

RUN chmod a+x /usr/local/bin/report.R</code></pre>
<p>Checkout the command line</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fl &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/system.file">system.file</a></span>(<span class="st">"docker/reporttool/src/report.R"</span>,
                  <span class="dt">package =</span> <span class="st">"sevenbriges"</span>)
<span class="kw">cat</span>(<span class="kw">readLines</span>(fl), <span class="dt">sep=</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</code></pre></div>
<p>Checkout the tool generator</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fl &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/system.file">system.file</a></span>(<span class="st">"docker/reporttool/rabix/generator.R"</span>,
                  <span class="dt">package =</span> <span class="st">"sevenbriges"</span>)
<span class="kw">cat</span>(<span class="kw">readLines</span>(fl), <span class="dt">sep=</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</code></pre></div>
</div>
</div>
<div id="compose-a-full-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#compose-a-full-workflow" class="anchor"></a>Compose a full workflow</h2>
<p>Even though in R with <code>sevenbridges</code> package you can use <code>%&gt;&gt;%</code> to connect two Tool object, but this only works for linear simple flow. For a complicated flow, we recommend you to use graphic user interface, it’s a lot of fun.</p>
<p>Now connect your RNA-seq tool with the report tool you just add it to your project, follow the tutorial <a href="http://docs.cancergenomicscloud.org/docs/the-pipeline-editor">here</a>, then make a workflow like this:</p>
<p><img src="https://i.imgur.com/dTKvLdt.png" width="100%" class="img-responsive" alt="cgc-workflow"></p>
<p>And next run a task on the platform via UI like this:</p>
<p><img src="https://i.imgur.com/6VsC69W.png" width="100%" class="img-responsive" alt="rnaseq-runtask"></p>
</div>
</div>
<div id="exercise-bring-your-own-tool" class="section level1">
<h1 class="hasAnchor">
<a href="#exercise-bring-your-own-tool" class="anchor"></a>Exercise: bring your own tool</h1>
<p>Now use either graphic user interface or R to describe your tool, your workflow, write report template for your tool (hard coded) and then share your flow with your friends.</p>
</div>
<div id="more-tutorials" class="section level1">
<h1 class="hasAnchor">
<a href="#more-tutorials" class="anchor"></a>More tutorials</h1>
<p>After you installed the package</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">browseVignettes</span>(<span class="st">"sevenbridges"</span>)</code></pre></div>
<p>or on the <a href="https://sbg.github.io/sevenbridges-r/articles/index.html">vignettes page</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li>
<a href="#prerequisites">Prerequisites</a><ul class="nav nav-pills nav-stacked">
<li><a href="#installation">Installation</a></li>
      <li><a href="#register-on-cancer-genomics-cloud">Register on Cancer Genomics Cloud</a></li>
      <li><a href="#authentication">Authentication</a></li>
      <li><a href="#register-on-shinyapps-io-optional">Register on shinyapps.io (Optional)</a></li>
      <li><a href="#report-issues">Report issues</a></li>
      </ul>
</li>
      <li>
<a href="#quickstart">Quickstart</a><ul class="nav nav-pills nav-stacked">
<li><a href="#create-a-project-under-your-account-via-api-r-client">Create a project under your account via API R client</a></li>
      <li><a href="#build-a-rna-seq-tool-from-bam-to-report">Build a RNA-Seq tool: from bam to report</a></li>
      <li><a href="#build-a-reporting-tool">Build a reporting Tool</a></li>
      <li><a href="#compose-a-full-workflow">Compose a full workflow</a></li>
      </ul>
</li>
      <li><a href="#exercise-bring-your-own-tool">Exercise: bring your own tool</a></li>
      <li><a href="#more-tutorials">More tutorials</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Nan Xiao, Dusan Randjelovic, Tengfei Yin, Seven Bridges Genomics.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
